\documentclass{bioinfo}
\copyrightyear{0000}
\pubyear{0000}

%%
%% Process using latex, since I use ps graphics (pdflatex requires bitmap).
%%
%% To get the \textcolor command to display, I processed using regular
%% latex.  The dvi file does not show the red text, but after
%% converting to either postscript (using dvips) or pdf (using
%% dvipdf), the color shows.
%%
%% For multiple footnotes.  Usage:
%%
%%     \footnoteremember{myfootnote}{This is my footnote}
%%
%% and then refer to this footnote again with
%%
%%     \footnoterecall{myfootnote}.
\newcommand{\footnoteremember}[2]{
  \footnote{#2} \newcounter{#1} \setcounter{#1}{\value{footnote}}}

\newcommand{\footnoterecall}[1]{\footnotemark[\value{#1}]}


\begin{document}
\firstpage{1}

\title[Unsupervised segmentation of continuous genomic data]{Unsupervised
  segmentation of continuous genomic data}
\author[Day \textit{et~al}]{Nathan Day\,$^{\rm
  a,}$\footnoteremember{foot1}{These authors contributed equally.}, 
Andrew Hemmaplardh\,$^{\rm a,}$\footnoterecall{foot1},
Robert E. Thurman\,$^{\rm b,c}$, John A. Stamatoyannopoulos\,$^{\rm c}$ and
William S. Noble\,$^{\rm a,b,c,}$\footnote{To whom correspondence should be addressed}}
\address{
$^{\rm a}$Department of Computer Science and Engineering,
$^{\rm b}$Division of Medical Genetics,
$^{\rm c}$Department of Genome Sciences, University of Washington,
  Seattle, WA, USA
}
\maketitle

\begin{abstract}
\section{Summary:}
The advent of high-density, high-volume genomic data has created the
need for tools to summarize large datasets at multiple scales.  HMMSeg
is a command-line utility for the scale-specific segmentation of
continuous genomic data using hidden Markov models (HMMs).
Scale-specificity is achieved by an optional wavelet-based smoothing
operation.  HMMSeg is capable of handling multiple datasets
simultaneously, rendering it ideal for integrative analysis of
\textcolor{red}{expression}, phylogenetic, and functional genomic
data.
\section{Availability:}
\href{http://noble.gs.washington.edu/proj/hmmseg}{http://noble.gs.washington.edu/proj/hmmseg}
\section{Contact:} \href{noble@gs.washington.edu}{noble@gs.washington.edu}
\end{abstract}

\section{Introduction}

The convergence of the genomic era and the advent of
high-throughput biological and chemical assays has created a wealth of
genomic data, much of which is presented in continuous,
time-series-like fashion across the genome.  Often it is desirable to
extract simplifying summary information from such data.  One
summarization approach involves segmenting the data into a small
number of discrete states based on the continuous output values.  This
segmentation may be accomplished in an unsupervised fashion using
hidden Markov models (HMMs).  For example, a chromosome-wide
continuous profile of bulk RNA output generated using tiling DNA
microarrays may be partitioned by segmenting the chromosomal
coordinates into three states, corresponding to regions of low, medium
and high transcription levels.  This type of categorization is often
desirable in the context of elucidating broad, large-scale trends in
the data.  In this case, it may be preferable to smooth the data to a
specified scale before segmentation, in order to eliminate spurious
state transitions resulting from isolated fine-scale features (see
Figure \ref{scales}).

HMMSeg is a tool for segmenting continuous genomic datasets on a
scale-specific basis using HMMs.  Scale specificity is achieved by an
optional smoothing step using wavelets (see below).  HMMSeg provides
multivariate capability, computing a single segmentation based on
multiple datasets simultaneously defined on a common set of genomic
coordinates.  

As a platform for segmenting a wide variety of genomic data, HMMSeg is
distinguished from existing programs using HMMs, which typically fall
under two categories: toolboxes for applications in any field, such as
htk \citep{young:htk} \textcolor{red}{and GHMM (http://ghmm.org)}; or
biological application-specific tools that use HMMs, such as
glimmerHMM \citep{majoros:tigrscan}, for gene finding, and HMMer
\citep{eddy:hmmer}, for sequence analysis.


\begin{figure}
  \centering
  \includegraphics[width=2.6in]{Images/dummy.eps}
  \caption{Effect of wavelet smoothing on HMM segmentations.  (a)
  histone modification H3K4me1 (avg. raw data resolution $\sim900$
  bp).  (b) two-state segmentation of the data in (a), with black and
  white bars representing the two different states. (c) two-state
  segmentation of data in (a) following 20kb wavelet-smoothing. (Data
  excerpted from \citep{thurman:identification}.)}
  \label{scales}
\end{figure}

\subsubsection*{Hidden Markov Models}
A hidden Markov model (HMM) is a statistical model in which data are
assumed to be generated by a stochastic process defined by a
predetermined number of hidden states \citep{rabiner:tutorial}.  Each
state is defined by an {\em emission distribution}, from which data
values are generated.  The model also specifies probabilities for
transitioning between states.  The parameters defining the emission
and transition probabilities are typically learned from the data by
expectation maximization (EM).  Given the learned parameters, there
are two common methods for determining the state labels for each
observation: the Viterbi algorithm, which finds the single most
probable path (sequence of states); and posterior decoding, which
computes the most likely state at each point of the sequence.  HMMSeg
uses Gaussian emission distributions\textcolor{red}{, with diagonal
covariance for multiple datasets (assuming independence between
variables}), and supports both the Viterbi and posterior decoding
methods for state assignments.

\subsubsection*{Wavelet Smoothing}
Wavelets are a mathematical tool for multi-scale analysis
\citep{percival:wavelet}.  Though first used in practical applications
in the fields of engineering and signal processing, in recent years
wavelets have found many applications in computational biology
\citep{lio:wavelets}.  We apply scale-specific smoothing using a
variant of the discrete wavelet transform (DWT) called the maximal
overlap discrete wavelet transform (MODWT) \citep{percival:wavelet}.
Both the DWT and MODWT can be used to decompose a given signal via
{\it multiresolution analysis} into a sum of scale-specific signals.
In contrast with other smoothing techniques, wavelet smoothing is
essentially the process of subtracting out the small-scale behavior
rather than averaging it. HMMSeg uses the LA(8) family of wavelets for
all wavelet transforms. \textcolor{red}{The choice of wavelet scale is
application-dependent, and can be informed by prior biological
information about the scale of features of interest, or by
trial-and-error to achieve, say, a desired segment length
distribution. See the website for further details on wavelet
smoothing.}

\begin{figure}
  \centering
  \includegraphics[width=3in]{Images/dummy.eps}
  \caption{Multi-datatype functional domains defined using HMMSeg.
   Shown in this 1.7Mb region on chr 21 (ENCODE region ENm005) are raw
   data (top) and 64kb smoothed data (bottom) for DNA replication
   timing (green), RNA transcription (blue), and histone modifications
   H3K27me3 (purple) and H3ac (orange).  Row nine shows a two-state
   Viterbi segmentation based on all four datasets, with active domains
   in black and inactive in white. Note the concentration of genes
    (bottom row) within active domains. (Data displayed in UCSC Genome
  Browser with colors added later.)}
  \label{ucsc}
\end{figure}

\section{Description of functionality}

HMMSeg provides a command-line interface.  The input to HMMSeg is one
or more collections of files in either single-column or tab-delimited
BED format.  Each collection represents a different dataset; multiple
collections trigger a multivariate segmentation.  Data are assumed to
be evenly spaced, a requirement of both the HMM and wavelet
processing.  After reading the data from the input files, HMMSeg
optionally smooths the data at a user-specified scale using the MODWT.
There is also an option to smooth the data without HMM training.

HMMSeg proceeds to train a completely connected HMM on the data by using
EM.  By default, the HMM has two states; models with more states may
also be specified.  The Gaussian parameters and transition
probabilities are initialized randomly, although the user may provide
model parameters to replace or initialize EM training.  Training may
be repeated multiple times from different random starts, in which case
the model with the highest total likelihood is 
selected.  Based on the final model, observations are assigned to
states using the Viterbi algorithm or posterior decoding.

The program outputs the trained model plus the state assignment for
each observation.  If the user provides input data in BED format, then
the segmentation is output in {\it wiggle} format, suitable for
display in the UCSC Genome Browser \citep{kent:human}.  The wiggle
file contains separate tracks for the original data, smoothed data,
the state assignments, and (for the posterior decoding method) the
probabilities of each data point belonging to each state.

HMMSeg is implemented in Java for platform independence.  It has been
successfully tested on Windows and UNIX-style
systems. \textcolor{red}{Validation and accuracy test results are
available on the website.}

\section{Example}

In a recent study \citep{thurman:identification} we analyzed a number
of independently generated experimental datasets produced under the
NHGRI ENCODE project \citep{encode:encode}, whose ultimate goal is to
identify all of the functional elements in the human genome.
Currently the ENCODE project is in its pilot phase, analyzing 44
regions spanning 30MB ($\sim 1\%$) of the genome
\citep{encode:encode2}.  Our aim was to integrate multiple functional
datatypes to create a functional domain map of the ENCODE regions.  We
used HMMSeg to segment the data at the 64kb scale into two states,
interpreted {\it a posteriori} as functionally `active' or `inactive'.
We successfully applied this technique to individual datatypes and up
to five datasets simultaneously. Examples of large-scale domains
delineated by HMMSeg are pictured in Figure~\ref{ucsc}
\textcolor{red}{(see website for details). Here we see the advantages
of using HMMs over simple thresholding techniques, the logic of which
breaks down in scenarios with multiple datasets and few states.}  This
approach highlights the potential for integrating multiple functional
genomic datatypes with widely varying experimental resolution.

\section*{Acknowledgment}

We thank Charles Grant and Tobias Mann for helpful advice. Funding:
NIH grants U01~HG003161 and R01~GM071923.

\bibliographystyle{plainnat}
\bibliography{refs} % Check out the 'refs' CVS module and add it to
		    % your Bibtex path.

\end{document}
