\documentclass{bioinfo}
\copyrightyear{0000}
\pubyear{0000}

%%
%% Process using latex, since I use ps graphics (pdflatex requires bitmap).
%%
%% To get the \textcolor command to display, I processed using regular
%% latex.  The dvi file does not show the red text, but after
%% converting to either postscript (using dvips) or pdf (using
%% dvipdf), the color shows.
%%
%% For multiple footnotes.  Usage:
%%
%%     \footnoteremember{myfootnote}{This is my footnote}
%%
%% and then refer to this footnote again with
%%
%%     \footnoterecall{myfootnote}.
\newcommand{\footnoteremember}[2]{
  \footnote{#2} \newcounter{#1} \setcounter{#1}{\value{footnote}}}

\newcommand{\footnoterecall}[1]{\footnotemark[\value{#1}]}


\begin{document}
\firstpage{1}

\title[Open-source peptide identification algorithm]{Open-source
peptide identification algorithm}
\author[Park \textit{et~al}]{Christopher Park$^{\rm a}$, Aaron Klammer$^{\rm a}$,
Lukas K\"{a}ll$^{\rm a}$, 
William S. Noble\,$^{\rm a,b,}$\footnote{To whom correspondence should be addressed}}
\address{
$^{\rm a}$Department of Computer Science and Engineering,
$^{\rm b}$Department of Genome Sciences, University of Washington,
  Seattle, WA, USA
}
\maketitle

\begin{abstract}
\section{Summary:}

We identify peptides better than anyone else! 

\section{Availability:}
\href{http://noble.gs.washington.edu/proj/crux}{http://noble.gs.washington.edu/proj/crux}
\section{Contact:} \href{noble@gs.washington.edu}{noble@gs.washington.edu}
\end{abstract}

\section{Introduction}

% Paragraph - Motivation   - Define the peptide identification problem
Peptide identification is hard!

ficulty in protein identification. In addition, the process is further complicated because each spectrum is a result of a cleaved peptide product, while the information we desire is the original source protein. Thus, the protein identification problem can be divided into two steps. First, the user must identify the peptide from the spectrum. Second, with the peptide spectrum match information the source proteins must be inferred. Majority of the effort in the field has been focused on improving peptide identification, because still many algorithms fail to accurately identify peptides in a timely manner.
	There are two major approaches for peptide identifications. First, the peptide database approach, where all candidate peptides are compared to the query spectrum in which the highest scoring peptide is identified. Second, the de novo sequencing approach, which tries to infer the most statistically significant sequence from the spectrum without any peptide database. In practice, the peptide database approach is most widely used because of the lack of accuracy in de novo sequencing. In database searching algorithms, a major bottleneck is the identification of candidate peptides within a given mass window. Mainly because the search space for a given proteome is quite large. For instance, the human proteome can result in 550 million peptides of length 6 to 50 amino acids.

% Paragraph - Related Work - SEQUEST, Tandem as open source?
SEQUEST\citep{eng:approach}
% Paragraph - Related Work - Indexing
% Paragraph - Related Work - Target decoy searches
% Paragraph - Related Work - Percolator
% Paragraph - Contribution - Four points of improvement
% Paragraph - Contribution - Brief results description


\begin{figure}
  \centering
  \includegraphics[width=2.6in]{Images/dummy.eps}
  \caption{Peptide indexing allows rapid generation of candidate peptide
  lists.}
  \label{figure:indexing}
\end{figure}

\begin{figure}
  \centering
  \begin{tabular}{c}
  \includegraphics[width=2.6in]{Images/dummy.eps} \\
  \includegraphics[width=2.6in]{Images/dummy.eps} \\
  \caption{Re-implementation of Sp and Xcorr scoring functions.}
  \label{figure:indexing}
  \end{tabular}
\end{figure}

\begin{figure}
  \centering
  \includegraphics[width=2.6in]{Images/dummy.eps}
  \caption{Positives vs. q-value (a measure of false discovery rate) for
  standard database search algorithms, and our implementation.}
  \label{figure:indexing}
\end{figure}



\section{Algorithms}

\subsection*{Indexing}
\subsection*{Scoring}
\subsection*{Null peptide generation}
\subsection*{Percolator}
\subsection*{Parallelization}

\section{Results}

We beat everybody!

\subsection*{Datasets}
We use the following publicly available tandem mass spectrometry data sets.

\section{Discussion}
% Mention Ting Chen's contribution
Discuss this!

\section*{Acknowledgment}

Acknowledge this!

% We thank Grant and Tobias Mann for helpful advice. Funding:
% NIH grants U01~HG003161 and R01~GM071923.


% N.b. to use the bibliography, check out the 'refs' CVS module and add it to
% your Bibtex path.

\bibliographystyle{plainnat}
\bibliography{refs} 

\end{document}
