\documentclass{article}

\usepackage[margin=1in]{geometry}

\begin{document}

\title{Improved candidate peptide generation using precomputed peptide
database}

\author{
Christopher Park\\
Department of Genome Sciences\\
University of Washington\\
Seattle, WA, USA
\and
Aaron Klammer\\
Department of Genome Sciences\\
University of Washington\\
Seattle, WA, USA
\and
William Stafford Noble\\
Department of Genome Sciences\\
Department of Computer Science and Engineering\\
University of Washington\\
Seattle, WA, USA
}

\maketitle

\section{Introduction}

In algorithms that are used to search sequence database with
tandemMSspectra data,  such as SEQUEST (Eng et al. 1995), one major
bottle neck is the identification of  candidate peptides within a
given mass window. Especially for larger proteomes the  search space
is quite large. For instance, the human proteome can result in 550
million  peptides from the range of 6 to 50 amino acids. We present an
efficient method for  generating peptides in a given mass window by
using a precomputed peptide  database. Our method reduces (some
day...) the runtime dramatically for large  proteome searches compared
to SEQUEST that does not use any precomputed  database. 

\section{Methods}

First a user creates a peptide database on disk for the proteome of
interest. During this  process, computed peptides get swapped in and
out from memory onto disk to prevent  memory shortages. Also, to
reduce the disk size of the database, instead of storing the  sequence
of each individual peptide, only the protein index, the start index
and the  peptide length are stored. This way the program can quickly
retrieve the sequence  from the original Fasta file. Once the peptide
database is on disk, for all queries given a mass window,  peptides
that meet the constraint are simply parsed out from the precomputed
peptide  database. This use of the peptide database eliminates the
need to recompute peptides  for each search.

\section{Preliminary Data}

We test our method by comparing it to stateofart technology
SEQUEST. First, the  runtime comparison was conducted by running a
single spectrum against E. coli and  Human proteome on both SEQUEST
and our method. SEQUEST uses a preliminary  scoring method called `Sp'
to extract the top 500 candidate peptides before the final
crosscorrelation analysis. Since we are unable to directly measure the
runtime for SEQUEST to generate its candidate peptides, we compare the
combined runtime of  generating a list of candidate peptides, and
performing Sp on all the peptides. We have  implemented Sp in our
method that is identically to SEQUEST. The results show that  our
method by using a precomputed peptide database outperforms SEQUEST
especially on large proteomes. The average runtime against E. coli for
single spectrum  for SEQUEST is XXXs, while our method is XXXs. For
the larger Human proteome,  SEQUEST is XXXs, while our method is XXXs.
Second, the accuracy comparison was conducted by comparing the peptide
list of  the top 500 candidate peptides (Sp ranked) generated by
SEQUEST and our method.  We compared 10 different spectra against
E. coli and Human proteome, in which our  method generated an
identical peptide list to SEQUEST. In addition, although  SEQUEST does
not return the entire peptides list, it does report the total matched
peptide count for each search. This peptide count was consistent with
our peptide  count for all 10 trial runs. Finally, the peptide
database required 34hrs of computation time and 17GB disk  size for
the Human proteome. The smaller E. coli proteome required 1hr40mins of
computation and 1.8GB of disk size.

\section{Novel Aspect}

When compared to current searching methods, our method is first to use
a pre computed peptide database that yields faster peptide searches
for large proteomes.

\end{document}

