\documentclass[12pt]{article}

\usepackage[margin=1in]{geometry}

\begin{document}

\title{Improved candidate peptide generation using a precomputed
peptide database}

\author{
Christopher Park\\
Department of Genome Sciences\\
University of Washington\\
Seattle, WA, USA
\and
Aaron Klammer\\
Department of Genome Sciences\\
University of Washington\\
Seattle, WA, USA
\and
William Stafford Noble\\
Department of Genome Sciences\\
Department of Computer Science and Engineering\\
University of Washington\\
Seattle, WA, USA
}

\maketitle

\section{Introduction}

In algorithms such as SEQUEST (Eng et al. 1995) that are used to
search sequence databases with tandem MS spectra data, one major
bottleneck is the identification of candidate peptides within a given
mass window. Especially for larger proteomes the search space is quite
large. For instance, the human proteome includes 550 million peptides
with lengths between 6 and 50 amino acids. We present an efficient
method for retrieving all peptides within a given mass window by using
a precomputed peptide database.  Compared to SEQUEST, our method
dramatically reduces the running time for large proteome searches.
The method is implemented in a publicly available software package
called ``Crux.''

\section{Methods}

Searching with Crux requires two phases.  First, a user creates a
peptide database on disk for the proteome of interest. During this
process, computed peptides get swapped in and out of memory onto disk
to prevent memory shortages. Also, to reduce the disk size of the
database, instead of storing the sequence of each individual peptide,
only the protein index, the start index and the peptide length are
stored.  Once the peptide database is stored on disk, retrieving all
peptides within a given mass window can be accomplished efficiently.
This use of the peptide database eliminates the need to re-scan the
peptide database for each search.

\section{Preliminary Data}

Using Crux, we precomputed a peptide database for the {\em E.\ coli}
and human proteomes.  For {\em E.\ coli} this procedure required 1 hr
40 mins of computation and 1.8GB of disk space.  For the human
proteome, the requirements were 34 hrs of computation time and 17GB of
disk space.

SEQUEST uses a preliminary score called ``Sp'' to extract the top 500
candidate peptides prior to the final cross-correlation analysis.  We
re-implemented the computation of this score, and we searched ten
spectra against the {\em E.\ coli} and human proteomes using SEQUEST
and using our software.  For each spectrum, we verified that SEQUEST
and Crux retrieve the same total number of candidate peptides, and
that the top 500 peptides (which are the only ones reported by
SEQUEST) are identical.

Finally, we compared the efficiency of SEQUEST and Crux for retrieving
and assigning preliminary scores to peptides.  We searched X spectra
against the {\em E.\ coli} and human proteomes using an m/z window of
X.  For {\em E.\ coli}, the average running time was XXX s for SEQUEST
and XXX s for Crux.  For the human proteome, the running times were
XXX s and XXX s.

\section{Novel Aspect}

When compared to current searching methods, our method uses a
precomputed peptide database that yields faster peptide searches for
large proteomes.

\end{document}

