<html>
<head>
<title>Crux tandem mass spectrometry analysis software</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<H1 ALIGN=CENTER>Crux tandem mass spectrometry analysis software</H1>
<center>
<img  src="schematic.png" usemap="#commands-files">
<map name="commands-files">
<area shape="rect" coords="169,92,304,123"  alt="create-index" href="crux-create-index.html" />
<area shape="rect" coords="169,188,297,218" alt="bullseye" href="bullseye.html " />
<area shape="rect" coords="498,104,634,133" alt="search-for-matches" href="crux-search-for-matches.html" />
<area shape="rect" coords="498,158,634,186" alt="search-for-xlinks" href="crux-search-for-xlinks.html" />
<area shape="rect" coords="822,71,954,100"  alt="barista" href="barista.html " >
<area shape="rect" coords="822,108,950,133" alt="percolator" href="percolator.html " />
<area shape="rect" coords="822,144,952,171" alt="q-ranker" href="q-ranker.html" />
<area shape="rect" coords="822,181,952,206" alt="calibrate-scores" href="calibrate-scores.html " />
<area shape="rect" coords="822,216,953,242" alt="spectral-counts" href="spectral-counts.html" />

</map>

</center>

<p>
Crux is a software toolkit for tandem mass spectrometry analysis, with
a focus on peptide identification.  If you use Crux in your research,
please cite</p>

<blockquote>
Christopher Y. Park, Aaron A. Klammer, Lukas K&auml;ll, Michael
J. MacCoss and William Stafford
Noble.  <a href="http://noble.gs.washington.edu/papers/park2008rapid.html">"Rapid
and accurate peptide identification from tandem mass spectra."</a>
<i>Journal of Proteome Research.</i> 7(7):3022-3027, 2008.</li>
</blockquote>

<p>
Crux is provided as a single executable that supports the following
commands:</p>

<table>

<tr>
<td colspan="2"><b>Primary commands</b></td>
</tr>

<tr>
<td><a href="bullseye.html">bullseye</a></td>
<td>Assign high resolution precursor m/z values to MS/MS data using the Hardkl&ouml;r algorithm.
</tr>

<tr>
<td>
<a href="crux-create-index.html">create-index</a></td>
<td>
Speed up subsequent calls to <code>search-for-matches</code> by
creating an index for all peptides in a fasta file.</td>
</tr>

<tr>
<td>
<a href="tide-index.html">tide-index</a></td>
<td>
Speed up subsequent calls to <code>tide-search</code> by creating an
index for all peptides in a fasta file.</td>
</tr>

<tr>
<td>
<a href="comet.html">comet</a></td>
<td>
Search a collection of spectra against a sequence database, returning
a collection of peptide-spectrum matches (PSMs).  This is Jimmy Eng's
search engine.  It is the most robust and easiest to use of the three
search engines in Crux.</td>
</tr>

<tr>
<td>
<a href="crux-search-for-matches.html">search-for-matches</a></td>
<td>
Search a collection of spectra against a sequence database, returning
a collection of peptide-spectrum matches (PSMs).  This is the
original Crux search engine.  It offers some features not available
in the other two search engines.</td>
</tr>

<tr>
<td>
<a href="tide-search.html">tide-search</a></td>
<td>
Search a collection of spectra against a sequence database, returning
a collection of peptide-spectrum matches (PSMs).</td>
</tr>

<tr>
<td>
<a href="calibrate-scores.html">calibrate-scores</a></td>
<td>
Assign two types of statistical confidence measures (q-values and
posterior error probabilities) to each PSM in a given set.</td></tr>

<tr>
<td>
<a href="percolator.html">percolator</a></td>
<td>
Re-rank a collection of PSMs using the Percolator algorithm.</td></tr>

<tr>
<td>
<a href="q-ranker.html">q-ranker</a></td>
<td>
Re-rank a collection of PSMs using the Q-ranker algorithm.</td></tr>

<tr>
<td>
<a href="barista.html">barista</a></td>
<td>
Rank proteins and assign a confidence measure to each
protein identification.</td></tr> 


<tr>
<td>
<a href="crux-search-for-xlinks.html">search-for-xlinks</a></td>
<td>
Search a collection of spectra against a sequence database, finding
cross-linked peptide matches.</td></tr>

<tr>
<td>
<a href="spectral-counts.html">spectral-counts</a></td>
<td>
Quantify peptides or proteins using one of three spectral counting methods.</td></tr>

<tr>
<td colspan="2"><b>Utilities</b></td>
</tr>

<tr>
<td><a href="make-pin.html">make-pin</a></td>
<td>Given a set of search results files, generate a pin.xml file for input to
<code>crux percolator</code></td>
</tr>

<tr>
<td>
<a href="crux-predict-peptide-ions.html">predict-peptide-ions</a></td>
<td>
Given a peptide and a charge state, predict the m/z values of the
resulting fragment ions.</td>
</tr>

<tr>
<td><a href="hardklor.html">hardkl&ouml;r</a></td>
<td>Identify isotopic distributions from high-resolution mass spectra.</td>
</tr>

<tr>
<td>
<a href="crux-print-processed-spectra.html">print-processed-spectra</a></td>
<td>
Process spectra as for scoring xcorr and print the results to a file.</td></tr>

<tr>
<td>
<a href="generate-peptides.html">generate-peptides</a></td>
<td>
Extract from a given set of protein sequences a list of target and
decoy peptides fitting the specified criteria.</td>
</tr>

<tr>
<td>
<a href="crux-get-ms2-spectrum.html">get-ms2-spectrum</a></td>
<td>
Extract one or more MS/MS spectra, specified by scan number, from
an <code>MS2</code> file.</td>
</tr>

<tr>
<td>
version</td>
<td>
Print the Crux version number to standard output, then exit.</td>
</tr>

<tr>
<td colspan="2"><b>Utilities for processing tab-delimited text files</b>
</tr>

<tr>
<td>
<a href="extract-columns.html">extract-columns</a></td>
<td>Print specified columns from a tab-delimited file.</td>
</tr>

<tr>
<td>
<a href="extract-rows.html">extract-rows</a></td>
<td>Print specified rows from a tab-delimited file.</td>
</tr>

<tr>
<td>
<a href="stat-column.html">stat-column</a></td>
<td>Collect summary statistics from a column in a tab-delimited file.</td>
</tr>

<tr>
<td>
<a href="sort-by-column.html">sort-by-column</a></td>
<td>Sort a tab-delimited file by a column.</td>
</tr>

</table>

<p>
<b>File formats</b></p>

<ul>
<table border="1">
<tr>

<td>&nbsp;</td>

<td>
<a href="ms2-format.html">MS1</a>
</td>

<td>
<a href="ms2-format.html">MS2</a>
(<a href="data/demo.ms2">sample</a>)<br>
mzML (1.0 and 1.1), <br>mzXML, MGF, CMS2
</td>

<td>
<a href="fasta-format.html">FASTA</a><br>
(<a href="sample.fasta">sample</a>)</td>

<td>
database index</td>

<td>
<a href="txt-format.html">Tab-delimited text</a><br>
(<a href="data/search.target.txt">sample</a>)</td>

<td>
pepXML</td>

<td>
Percolator XML<br> (<a href="https://github.com/percolator/percolator/blob/master/src/xml/percolator_in.xsd">PIN</a> / <a href="https://github.com/percolator/percolator/blob/master/src/xml/percolator_out.xsd">POUT</a>)
</td>

<td>mzIdentML</td>

<td>
<a href="sqt-format.html">SQT</a>
</td>

<td>
<a href="features.html">Features</a>
</td>

<td>
<a href="barista-xml-format.html">Barista XML</a>
</td>


<td>
<a href="parameter-file.html">Crux parameters</a><br>
(<a href="default.params">sample</a>)</td>
</tr>

<tr bgcolor="#DEDEDE">
<td>bullseye</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>&nbsp;</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

<tr>
<td>create-index</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

<tr bgcolor="#DEDEDE">
<td>comet</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>&nbsp;</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>out</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

<tr bgcolor="#DEDEDE">
<td>search-for-matches</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>out</td>
<td>out</td>
<td>out</td>
<td>out</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

<tr>
<td>calibrate-scores</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
<td>in/out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

<tr bgcolor="#DEDEDE">
<td>percolator</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
<td>in/out</td>
<td>in/out</td>
<td>out</td>
<td>in</td>
<td>in/out</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

<tr>
<td>q-ranker</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in</td>
<td>out</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

<tr bgcolor="#DEDEDE">
<td>barista</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in</td>
<td>&nbsp;</td>
<td>out</td>
<td>in/out</td>
</tr>

<tr>
<td>search-for-xlinks</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>in</td>
<td>out</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

<tr bgcolor="#DEDEDE">
<td>spectral-counts</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
<td>in</td>
<td>&nbsp;</td>
<td>in</td>
<td>in</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>in/out</td>
</tr>

</table>
</ul>

<p>
<b>Tutorials</b>

<ul>

<li>
<a href="install-tutorial.html">Installation</a></li>

<li>
<a href="start-tutorial.html">Getting started with Crux</a></li>

<li>
<a href="search-tutorial.html">Searching a protein database</a></li>

<li>
<a href="index-tutorial.html">Creating a peptide index</a></li>

<li>
<a href="customize-tutorial.html">Customizing a search</a></li>

<li>
<a href="spectral-counts-tutorial.html">Using spectral-counts</a></li>

</ul>


<p>
<b>Download</b>

<ul>
Crux source code and binaries can be
downloaded <a href="download/index.html">here</a>. 
Crux is covered by an <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache
license</a>.
</ul>


<p>
<b><a href="crux-faq.html">Crux frequently asked questions</a></b></p>

<p>
<a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Crux Google Scholar profile</a></p>

<p>
To receive announcements of new versions, sign up for the Crux users 
<a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users">mailing list.</a></p>
<p>For support, and to discuss Crux with other users and the developers, 
sign up for the Crux users 
<a href="http://groups.google.com/group/crux-users">Google group.</a></p>
</p>

<p>
<a href="release-notes.html">Release notes.</a></p>

<p>
<a href="known-issues.html">Known issues.</a></p>
<hr></hr>

<center>
<p>
The original version of Crux was written by Chris Park and Aaron Klammer
under the supervision
of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
MacCoss</a>
and <a href="http://noble.gs.washington.edu/~noble">Prof. William
Stafford Noble</a> in the Department of Genome Sciences at the
University of Washington, Seattle.  The complete list of contributors
can be found <a href="contributors.html">here</a>.</p>

<p>Maintenance and development of Crux is funded by the National
Institute of General Medical Sciences grants R01&nbsp;GM096306
and P41&nbsp;GM103533.</p>
</center>

<hr></hr>

<p>
Please send comments and questions to cegrant@uw.edu.</p>

</blockquote>
</body>
</html>
