<html>
<head>
<title>Crux Frequently Asked Questions</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<h1 align=center>Crux Frequently Asked Questions</h1>

<ol>

<li>
<b>What input file formats does Crux support?</b></li>

<p>
Crux can search spectra in .ms2 and .mzXML file formats.  The .ms2
files can be generated from Thermo .RAW files using the <a 
href="http://proteome.gs.washington.edu/software.html">MakeMS2</a>
program.  The .ms2 format is described <a href="ms2-format.html">here</a>. 
When searching .mzXML files, use the <code>use-mstoolkit=T</code>
option in the parameter file.  Additional file formats will be
supported in future releases. 
</p>

<li>
<b>How does Crux differ from SEQUEST?</b></li>

<p>
SEQUEST has split into at least four different development threads:
(1) the version developed in the Yates lab, initially at the
University of Washington and later at Scripps, (2) the version mainted
by Thermo Scientific, (3) the version mainted by Sage-N, and (4) the
version maintained in the MacCoss lab at the University of Washington.
Crux is based on the Yates lab development thread, SEQUEST version 28,
whose last documented modification is dated 8/16/99.</p>

<p>
The Crux command <code>sequest-search</code> is an imperfect
reimplementation of SEQUEST 28.  The imperfection arose because, to
avoid copyright infringement, Crux was reverse engineered from the
SEQUEST binary; i.e., the Crux developers did not have access to the
SEQUEST source code.  The reverse engineering continued until the
developers deemed the remaining differences to be of negligible
importance.  A quantitative comparison of the scores computed by Crux
and SEQUEST can be seen in
the <a href="http://noble.gs.washington.edu/papers/park2008rapid.html">original
Crux paper</a>.  The differences between Crux and SEQUEST v28 are
comparable in magnitude to the differences between different versions
of SEQUEST.</p>

<li>
<b>How does Crux compute the masses of peptides and peptide
  fragments?</b></li>

<p>
<ul>

<li>
Each amino acid can be characterized either by its monoisotopic mass,
which is the mass of the most common isotopic form of that amino acid,
or the average mass, which is a weighted average of the masses of all
the isotopic forms.  By default, Crux uses average mass to calculate
peptide masses, though this can be controlled by the
<code>--isotopic-mass</code> option to <code>crux create-index</code>.
For fragments, Crux always uses the monoisotopic mass.</li>

<li>
The neutral mass of a peptide is not the sum of the masses of its
amino acids.  The C-terminus (i.e., the suffix, or y-ion) of the
peptide has an additional water group, whose mass is either
18.010564684 Da (monoisotopic) or 18.0153 Da (average).  Note that,
for cleavage ions, this additional mass should be added to the mass of
a y-ion, but not to a b-ion.</li>

<li>
By default, Crux adds a static modification of 57.021464 Da to all
cysteines.  This is because in most protein preparation protocols
the peptides are alkylated with iodoacetamise, resulting in 
carbidomethylation of cysteine. The alkylation step prevents the 
reformation of disulfide bonds.  Other alykylation reagents may be
used, in which case the appropriate mass shift can be specified with the
<code>C=[mass]</code> option.  Set <code>C=0</code> for no
modification.
</li>

<li>
A charged peptide has an additional approximately 1 Da mass for each
charge, corresponding to the mass of a hydrogen atom.  The exact mass
to be added depend on whether we are using the monoisotopic mass
(1.0078246) or average mass (1.00794).</li>

</ul>
</p>

<!--
<li>
<b>What is the difference between Crux Percolator and the stand-alone
version available
from <a href="http://www.per-colator.com">http://www.per-colator.com</a>?</b></li>
-->

<li>
<b>What operating systems does Crux work with?</b></li>

<p>
Crux is written in C++, so in principle it should work on virtually
any modern operating system.  We provide pre-compiled binaries for
Linux and Windows.  The latter requires that you have Cygwin
installed.</p>

<li>
<b>How does Crux select candidate peptides from the database?</b></li>

<p>
Crux selects candidate peptides for each spectrum based on its
precursor m/z and an assumed charge state (specified in the input
file).  A window around that precursor m/z is defined in one of three
ways based on the <code>precursor-window-type</code>
and <code>precursor-window</code> options.
<ul>
<li>  
If the <code>precursor-window-type</code> is set to <code>mz</code>,
then the window is calculated as precursor m/z &plusmn;
<code>precursor-window</code> and the resulting range is converted
to mass using the charge state.</li>
<li>
If the <code>precursor-window-type</code> is set to <code>mass</code>, then
the charge state is used to convert the precursor m/z to mass, and the
window is defined as precursor mass &plusmn; precursor-window.</li>
<li>
If the <code>precursor-window-type</code> is set to <code>ppm</code>
(parts per million), then the precursor m/z is converted to mass and
the window spans from precursor mass / (1.0 + window * 1e-6) to 
precursor mass / (1.0 - window * 1e-6).</li>
</ul>
Candidate peptides are those whose mass falls within the defined
window.  The peptide mass is computed as the sum of the average amino
acid masses plus 18 Da for the terminal OH group, plus 1 Da for each
charge.  Candidate peptides can also be constrained by minimum and
maximum allowed length. </p>

<p>
Peptides are selected from protein sequence given either in a fasta
file or a fasta file that has been indexed by <a href="crux-create-index.html">
<code>crux create-index</code></a>.  The options <code>enzyme</code>, 
<code>digestion</code> and <code>missed-cleavages</code> define how
the proteins are broken up into peptides.  When <code>enzyme</code> is
set to <code>no-enzyme</code> then any subsequence of a protein may be
considered as a candidate peptide, and the options <code>digestion</code>
and <code>missed-cleavages</code> are not used.  For other values of
<code>enzyme</code>, the residues at the termini of the protein
subsequence must follow specific rules.  For example, trypsin requires
that the preceeding residue must be an R or K and the following
residue may not be a P.  When <code>digestion</code> is set
to <code>full-digest</code>, then these rules must be true at both ends of the
peptide.  When it is <code>partial-digest</code>, then the rules must be true
at at least one of the ends.  If <code>missed-cleavages</code> is set
to false, then no cleavage site may lie within the peptide sequence;
alternatively, if true then any number of sites may lie within the
peptide sequence. </p>

<!--
<li>
<b>How is the theoretical spectrum for a peptide generated?</b></li>

<li>
<b>In the calculation of XCorr, what preprocessing steps does the
observed spectrum undergo prior to comparison with the theoretical
spectrum?</b></li>
-->

<li>
<b>What enhancements to Crux are planned?</b></li>

<p>
Crux is under constant development, so the following list of
enhancements may be out of date.  As of November, 2010, the list
includes the following:</p>

<p>
<ul>

<li>
Support for more input file types.</li>

<li>
The ability to exploit accurate precursor masses.</li>

<li>
The ability to compute peptide-level q-values
via <code>compute-q-values</code>, <code>percolator</code>
or <code>q-ranker</code>.</li>

<li>
Support for searching for post-translational modifications in Tide
and <code>search-for-xlinks</code>.</li>

<li>
Tools to do protein-level inference, including identification and
spectral counting.</li>

</ul>
</p>

<li>
<b>How can I contribute to Crux?</b></li>

<p>
Please email crux-support@uw.edu to obtain read-only access to our
Subversion repository, along with access to our development team web
pages.  Patches implementing new features can then be emailed to the
development team for review and inclusion in subsequent releases of
Crux.</p>

<li>
<b>Where does the name "Crux" come from?</b></li>

<p>
Thin air.  The name is not an acronym or a reference to anything in
particular.</p>

</ol>

</blockquote>
</body>
</html>


