<html>
<title>Crux tandem mass spectrometry analysis software</title>
<body>
<blockquote>

<H1 ALIGN=CENTER>Crux tandem mass spectrometry analysis software</H1>

<p>
One of the major challenges in proteomics is to identify the proteins
that are expressed in cells under different conditions.  Mass
spectrometry is a widely used method for protein identification;
however, a detailed spectrum analysis is required to match the
spectrum to the source protein. Crux contains software for tandem mass
spectrometry analysis, with a focus on peptide identification.
It can be downloaded <a href="download/index.html">here</a> once you agree to the
license agreement and obtain a password <a href="http://depts.washington.edu/ventures/UW_Technology/Express_Licenses/crux.php">here</a>.
Crux source code is freely available for academic use and can be
licensed for commercial use. Crux optionally 
makes use of the Percolator algorithm (v. 1.05), 
which can be downloaded separately <a
  href="http://noble.gs.washington.edu/proj/percolator/">here</a>.</li>

</p> 

<p>
The core programs in the Crux toolkit are</p>

<ul>

<li>
<a href="crux-search-for-matches.html">crux-search-for-matches</a>
Search a collection of spectra against a sequence database, returning
a collection of peptide-spectrum matches (PSMs).</li>

<li>
<a href="crux-create-index.html">crux-create-index</a> Create an index
for all peptides in a fasta file.</li>

<li>
<a href="crux-analyze-matches.html">crux-analyze-matches</a> Analyze
a collection of PSMs using one the Crux supported post-processing algorithms.</ul>

<p>
In addition, Crux contains the following utilities:</p>

<ul>

<li>
<a href="crux-predict-peptide-ions.html">crux-predict-peptide-ions</a>
Given a peptide and a charge state, predict the ions according to the
provided options. 

<li>
<a href="crux-generate-peptides.html">crux-generate-peptides</a>
Given protein sequencs in a fasta file return a list of peptides
fitting the specified criteria.

<li>
<a href="crux-get-ms2-spectrum.html">crux-get-ms2-spectrum</a> Extract
an MS-MS spectrum with a specified scan number from a
<code>.ms2</code> file.</li>

<li>
<a href="spectacle.html">spectacle</a> Create an image file of an
annotated spectrum.</li>

</ul>

<p>A tutorial for installing and using the above programs can be found <a href="home-tutorial.html">here</a>.</p>

<p>Crux is written in C by Chris Park and Aaron Klammer under
the supervision of <a
href="http://noble.gs.washington.edu/~noble">Prof. William Stafford
Noble</a> in the Department of Genome Sciences at the University of
Washington, Seattle. Development of Crux is supported by National Institutes of Health award R01 EB007057.</p>

</blockquote>
</body>
</html>
