<html>
<title>Crux tandem mass spectrometry analysis software</title>
<body>
<blockquote>

<H1 ALIGN=CENTER>Crux tandem mass spectrometry analysis software</H1>

<p>
One of the major challenges in proteomics is to identify the proteins
that are expressed in cells under different conditions.  Mass
spectrometry is a widely used method for protein identification;
however, a detailed spectrum analysis is required to match the
spectrum to the source protein. Crux contains software for tandem mass
spectrometry analysis, with a focus on peptide identification.</p>

<p>
The core programs in the Crux toolkit are</p>

<ul>

<li>
<a href="crux-search-for-matches.html">crux-search-for-matches</a>:
Search a collection of spectra against a sequence database, returning
a collection of peptide-spectrum matches (PSMs).</li>

<li>
<a href="crux-analyze-matches.html">crux-analyze-matches</a>: Analyze
a collection of PSMs using Percolator.</li>

</ul>

<p>
In addition, Crux contains the following utilities:</p>

<ul>

<li>
<a href="crux-create-index.html">crux-create-index</a> Create an index
for all peptides in a fasta file.</li>

<li>
<a href="crux-predict-peptide-ions.html">crux-predict-peptide-ions</a>
Given a peptide and a charge state, predict the ions according to the
provided options. 

<li>
<a href="crux-score-peptides-spectrum.html">crux-score-peptide-spectrum</a>
Score a single spectrum for matches to a list of peptides.</li>

<li>
<a href="crux-get-ms2-spectrum.html">crux-get-ms2-spectrum</a> Extract
an MS-MS spectrum with a specified scan number from a
<code>.ms2</code> file.</li>

<li>
<a href="spectacle.html">spectacle</a> Create an image file of an
annotated spectrum.</li>

</ul>

<p>The current version of Crux is 1.0</p>

<p>Crux is written in ANSI C for Unix/Linux platforms.</p>

<p>The Crux software was written by Chris Park and Aaron Klammer under
the supervision of <a
href="http://noble.gs.washington.edu/~noble">Prof. William Stafford
Noble</a> in the Department of Genome Sciences at the University of
Washington, Seattle.</p>

<p>
Crux source code is freely available for academic use and can be
licensed for commercial use.  To obtain a copy of Crux, please contact
Alan Yen at <a href="http://depts.washington.edu/ventures">University
of Washington Digital Ventures</a>.</p>

<p>
Crux is supported by National Institutes of Health award R01
EB007057.</p>

</blockquote>
</body>
</html>
