<HTML>
<HEAD>
<TITLE>crux-generate-peptides</TITLE>
</HEAD>
<BODY>
<blockquote>
<H1 ALIGN=CENTER>crux-generate-peptides</H1>
<B>Usage:</B>

<blockquote>
<code>crux-generate-peptides [options] &lt;protein&nbsp;input&gt;</code>
</blockquote>

<B>Description:</B>

<p>
Given a protein database as input, generate a list of 
peptides in the database that meet certain criteria (e.g. mass,
length) as output. The user can specify whether to use a 
pre-computed peptide index database or a text fasta file. </p>

<p>
<B>Input:</B>

<UL>
  <LI>&lt;protein input&gt; &ndash; 
  The name of the file in fasta format or the directory containing the
  protein index from which to parse proteins.
	
    </UL>
<B>Output:</B>

<UL>

  The program prints to standard output a series of header lines,
  describing how the peptides were generated, and according to what
  criteria, as follows:

  <pre>
  # PROTEIN DATABASE: &lt;protein input filename&gt;
  # OPTIONS:  &lt;options used to generate the file&gt;</pre>

  Then the program prints a series of tab delimited lines with the
  following fields, where [ ] indicates an optional field:

  <pre>
  &lt;mass&gt; &lt;protein-id&gt; &lt;peptide-start&gt; &lt;peptide-length&gt; [peptide-sequence]
  </pre>
	
</UL>
<B>Options:</B>

<UL>
<LI>
<code>--output-sequence &lt;T|F&gt;</code> &ndash; 
Include the peptide sequence in the output.  Default = F.
	
<!--
<LI>
<code>--output-trypticity &lt;T|F&gt;</code> 
Include the peptide trypticiy in output.  When cleavages=partial all
peptides are labeled PARTIAL even if fully tryptic. Default = F. 
-->

<LI>
<code>--sort &ltp;mass|length|lexical |none&gt;</code> &ndash; 
Specify the order in which
peptides are printed to standard output.  Default = none.

<LI>
<code>--min-mass &gt;float&lt;</code> &ndash; The minimum neutral mass of the
peptides to output. Default = 200.
	
<LI>
<code>--max-mass &gt;float&lt;</code> &ndash; The maximum neutral mass of the
peptides to output. Default = 7200.
	
<LI>
<code>--min-length &lt;int&gt;</code> &ndash; 
The minimum length of the peptides to output. Default = 6.
 	
<LI>
<code>--max-length &lt;int&gt;</code> &ndash; 
The maximum length of the peptides to output. Default = 50.
	
<LI>
<code>enzyme &lt;trypsin|chymotrypsin|elastase|clostripain|cyanogen-bromide|idosobenzoate|proline-endopeptidase|staph-protease|modified-chymotrypsin|elastase-trypsin-chymotrypsin|no-enzyme&gt; 
</code> &ndash; 
Enzyme to use for <i>in silico</i> digestion of protein sequences. Used
in conjunction with the options digestion and missed-cleavages. Use
'no-enzyme' for non-specific digestion.  Digestion rules are as
follows: enzyme name [cuts after one of these residues]|{but not before
one of these residues}.  trypsin [RK]|{P}, elastase [ALIV]|{P},
chymotrypsin [FWY]|{P}, clostripain [R]|[], cyanogen-bromide [M]|[],
iodosobenzoate [W]|[], proline-endopeptidase [P]|[], staph-protease
[E]|[], modified-chymotrypsin [FWYL]|{P},
elastase-trypsin-chymotrypsin [ALIVKRWFY]|{P},aspn []|[D] (cuts before
D). Default = trypsin.

<li>
<code>custom-enzyme &lt;residues before cleavage | residues after cleavage
&gt;</code> &ndash
Specify rules for <i>in silico</i> digestion of protein sequences.
Overrides the <code>enzyme</code> option.  Two lists of residues are
given enclosed in square brackets or curly braces and separated by a
|. The first list contains residues required/prohibited before the
cleavage site and the second list is residues after the cleavage site.  
If the residues are required for digestion, they are in square
brackets, '[' and ']'.  If the residues prevent digestion, then they
are enclosed in curly braces, '{' and '}'.  Use X to indicate all
residues.  For example, trypsin cuts after R or K but not before P
which is represented as <code>[RK]|{P}</code>.  AspN cuts after any
residue but only before D which is represented as <code>[X]|[D]</code>.

<LI>
<code>--digestion &lt;full-digest|partial-digest&gt;
Degree of digestion used to generate peptides (full-digest,
partial-digest). Either both ends or one end of a peptide must
conform to enzyme specificity rules. Default full-digest., Used in
conjunction with enzyme option when enzyme is not set to to
'no-enzyme'.  


<LI>
<code>--missed-cleavages &lt;T|F&gt;</code> &ndash; 
Allow missed cleavage sites within a peptide.  When used with
enzyme is specified; includes peptides containing one or more
potential cleavage sites. Default = F. 

<LI>
<code>--isotopic-mass &lt;average|mono&gt;</code> &ndash; 
Specify the type of isotopic masses to use when calculating the
peptide mass. Default = average. 

<li>
<code>--unique-peptides &lt;T|F&gt;</code> &ndash; 
For peptides appearing in multiple proteins, store a reference to only
one of those proteins.  Default = F.

<LI><code>--verbosity&lt;int&gt;</code> &ndash; Specify the verbosity of the current processes from 0-100.  
By default, verbosity is set to 100. 

<LI>
<code>--use-index &lt;T|F&gt;</code> &ndash; Specify whether a pre-computed on-disk
index should be used for retrieving the peptides instead of using a
fasta file.

<LI><code>--parameter-file &lt;filename&gt;</code> &ndash; A file containing
command-line or additional parameters.  See the <a
href="parameter-file.html">parameter documentation</a> page for
details.

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; Specify the verbosity of the
current processes. Each level prints the following messages, including
all those at lower verbosity levels: 0-fatal errors, 10-non-fatal
errors, 20-warnings, 30-information on the progress of execution,
40-more progress information, 50-debug info, 60-detailed debug info.
Default = 30.

</UL>

<B>Parameter file options:</B>
<ul>

<li>
<code>mod &lt;mass change&gt;:&lt;aa list&gt;:&lt;max per peptide&gt;
   &ndash; </code>
Consider modifications on any amino acid in aa list with at most
max-per-peptide in one peptide.  This parameter may be included
with different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11. 

<li>
<code>cmod &lt;mass change&gt;:&lt;max distance from protein C-terminus&gt;
&ndash;</code> 
Consider modifications on the C-terminus of any peptide whose
C-terminus is no more than max-distance residues from the protein
C-terminus.  Use -1 to consider the C-terminus all peptides regardless
of position in the protein.  This parameter may be included with
different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11. The same modifications must be given for any
post-search process (compute-q-values, q-ranker).

<li>
<code>nmod &lt;mass change&gt;:&lt;max distance from protein N-terminus&gt;
&ndash;</code> 
Consider modifications on the N-terminus of any peptide whose
N-terminus is no more than max-distance residues from the protein
N-terminus.  Use -1 to consider the N-terminus all peptides regardless
of position in the protein. This parameter may be included with
different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11. The same modifications must be given for any
post-search process (compute-q-values, q-ranker).

<li>
<code>max-mods &lt;n&gt; &ndash;</code>
The maximum number of modifications that can be applied to a single
peptide. Default = no limit.

<li>
<code>max-aas-modified &lt;n&gt; &ndash;</code>
The maximum number of modified amino acids that can appear in one
peptide.  Each aa can be modified multiple times. Default = no limit.

<li>
<code>&lt;A-Z&gt &lt;float&gt;</code> &ndash;
Specify static modifications.  This is a mass change applied to the
given amino acid (in single-letter-code A thru Z) for every peptide in
which it occurs.  Use the <code>mod</code> option for generating
peptides both with and without the mass change.  Default C=57.
</ul>

</ul>

<HR></hr>

<P>
<A HREF="crux.html">Crux home</A></P>

</blockquote>
</body>
</html>
