<html>
<head>
<title>crux search-for-matches</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<H1 ALIGN=CENTER>crux search-for-matches</H1>

<B>Usage:</B>

<blockquote>
<code>crux search-for-matches [options]
&lt;ms2&nbsp;input&nbsp;filename&gt; &lt;protein&nbsp;input&gt;
&lt;output&nbsp;directory&gt;
</code>
</blockquote>

<B>Description:</B>

<blockquote>
<p>
This command searches a protein database with a set of spectra.  For
each spectrum, candidate peptides are identified whose inferred m/z
lies within a specified range of the observed precursor m/z.  These
candidate peptides are first ranked with the SEQUEST Sp score.  The
top 500 matches are then re-scored using XCorr.  The input protein
database may either be in FASTA format or it may be a binary index
created by <code>crux create-index</code>.</p>

<p>
<b>Modifications:</b> Crux handles modifications as follows.  
The user specifies an allowed set of amino acid modifications, using
the options <code>mod</code>, <code>cmod</code> and <code>nmod</code>,
which are described below.  Before any search is performed, Crux
generates an exhaustive list of all possible combinations of amino
acid modifications that could be applied to a peptide.  For example,
if the user specifies a modification of +79 which can be present twice
on any peptide and a modification of +30 which can be present only
once , then the list will contain modifications corresponding to +79
on one residue, +79 on two different residues, +30 on one residue, +79
on one and +30 on one, +79 on two and +30 on one.  Subsequently, for
each spectrum, Crux performs one search for each possible combination of
modifications.  For example, if the precursor m/z for a spectrum is
800 da, the charge state is +2, and Crux is considering a modification
of +79, then Crux will retrieve from the database all candidate
peptides whose total mass is close to 221 da.  These candidates are
scored as usual, with a preliminary score and a final score, and the
top <i>n</i> candidate peptides are added to a composite, sorted list
of peptides.  Finally, after all modifications have been searched,
Crux reports for the current spectrum the top <i>m</i> peptides from
the composite list.</p>

</blockquote>

<p>
<B>Input:</B>

<UL>

<LI>
&lt;ms2 &gt; &ndash; The name of the file (in MS2 format) from which
  to parse the spectra. </li>

<LI>
&lt;protein input&gt; &ndash; 
  The name of the file in fasta format or the directory containing a
  protein index from which to retrieve proteins and peptides.</li>

<li>
&lt;output&nbsp;directory&gt; &ndash; The name of the directory to
place the output files.</li>

</UL>

<B>Output:</B>

<UL>
<p>
The program writes to the user-specified folder the following files:</p>
<ul>

<li>
<b>target.csm</b>: a binary file that contains the PSMs and additional
features.  This file can be used as input
to <code>crux&nbsp;compute-q-values</code>
or <code>crux&nbsp;percolator</code>.

<li>
<b>target.sqt</b>: an SQT file containing the PSMs

<li>
<b>target.txt</b>: a tab-delimited text file containing the PSMs.
Each line contains the following 14 columns: (1) first scan, (2) last
scan, (3) spectrum neutral mass, (4) charge, (5) protein id, (6) main
rank, (7) other rank, (8) peptide mass, (9) delta cn, (10) main score,
(11) other score, (12) b/y matched, (13) b/y total, (14) peptide
sequence</li>


<li>
<b>log.txt</b>: a log file containing a copy of all messages that were
printed to stderr.

</ul>

<p>
If decoys are enabled using <code>--number-decoy-set</code>, then
three files called binary.csm, binary.sqt and binary.txt are also
produced.</p>

<p>
Optionally, you may prepend each of the above filenames with a
specified string, using the <code>--fileroot</code> option.

</p>
</UL>

<B>Options:</B>
<UL>

<li>
<code>--fileroot &lt;string&gt;</code> &ndash; Prepend each file in 
the output directory with the specified string.</li>

<li>
<code>--use-index &lt;T|F&gt;</code> Use an index that has already
been created if true (T) or use a fasta file if false (F). When using
an index the 'protein input' argument will be the name of the
directory containing the index files (the same as the 'index name'
argument for crux-create-index).  Default = F.</li>

<li>
<code>--overwrite &lt;T|F&gt;</code> Replace existing files if true (T) or 
fail when trying to overwrite a file if false (F). Default =  F.</li>

<LI>
<code>--number-decoy-set &lt;0|1|2&gt;</code> &ndash; Specify the number
of decoy sets to generate for match_analysis. At least one decoy set
is required to run the algorithm 'percolator' in a subsiquent
crux-analyze-matches run.  Default = 2. </li>

<li>
<code>--compute-p-values &lt;T | F&gt; &ndash</code>
Estimate the paramters of the score distribution for each spectrum and
compute a p-value for each PSM.  This option must be used in order to
run <code>crux compute-q-values</code> on the results.</li>

<LI>
<code>--spectrum-min-mass &lt;float&gt;</code> &ndash; The lowest spectrum 
m/z to search in the ms2 file. Default = 0.0

<LI>
<code>--spectrum-max-mass &lt;float&gt;</code> &ndash; The highest spectrum 
m/z to search in the ms2 file. Default = no maximum.

<LI> 
<code>--spectrum-charge &lt;1|2|3|all&gt;</code> &ndash; The spectrum 
charges to search. With 'all' every spectrum will be searched and
spectra with multiple charge states will be searched once at each
charge state.  With 1, 2, or 3 only spectra with that charge will be
searched. Default = all. 

<LI><code>--parameter-file &lt;filename&gt;</code> &ndash; A file containing
command-line or additional parameters.  See the <a
href="parameter-file.html">parameter documentation</a> page for
details.

<LI>
<code>--write-parameter-file</code> &ndash; Create a file in the
output directory named <code>params.txt</code>.  This file contains
the name and value of all parameters/options for the current
search.  Not all parameters in the file may have been used in the
search.  The resulting file can be used with the <code>--parameter-file
option for other crux programs. </code>

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; Specify the verbosity of the
current processes. Each level prints the following messages, including
all those at lower verbosity levels: 0-fatal errors, 10-non-fatal
errors, 20-warnings, 30-information on the progress of execution,
40-more progress information, 50-debug info, 60-detailed debug info.
Default = 30.

<li>
<code>--version T</code> &ndash;
Print the version number and quit.  Please note that you must include
the 'T' after --version.

</UL>
<B>Parameter file options:</B>
<ul>
<li>
<code>fragment-mass &lt;average|mono&gt;</code> &ndash; Which isotopes to use
in calcuating fragment ion mass (average, mono). Default = mono.

<li>
<code>mass-window</code> &ndash;
Tolerance used for matching peptides to spectra.  Peptides must be
within +/- 'mass-window' of the spectrum mass.  Default = 3.0.

<li>
<code>ion-tolerance&lt;float&gt;</code> &ndash; Tolerance used for matching
observed peaks to predicted fragment ions.  Default = 0.5.

<li>
<code>max-rank-preliminary &lt;int&gt;</code> &ndash;  Number of psms per
spectrum to score after preliminary scoring.  Default = 500.

<li>
<code>top-match &lt;int&gt;</code> &ndash; The number of psms per
spectrum writen 
to the binary output file. Only relevant when
output-mode=&lt;all|binary&gt;. Does not affect output to text files.
If more than one psm is written to file, the matches for each spectrum
will be re-ranked in crux-match-analysis. Default = 1.

<li>
<code>top-match-sqt &lt;int&gt;</code> &ndash; 
The number of psms per spectrum written to the sqt file.  Only
relevant when output-mode=&lt;all|sqt&gt;.  Does not affect output to
binary files. Default = 5. 

<li>
<code>tdc &lt;T|F&gt;</code> &ndash;
Target-decoy competition.  Combine target and decoy searches in one
file.  Default FALSE, separate files. After searching target and decoy
peptides, all psms for one spectrum are sorted together and the mixed
list is returned in the target.sqt file. Set number of decoys with
num-decoys-per-target.  The number of psms retained after preliminary
searching (max-rank-preliminary) is increased to reflect the number of
decoys searched so that (1+num-decoys)*max-rank-preliminary are kept.  

<Li>
<code>num-decoys-per-target &lt;int&gt;</code> &ndash;
Number of decoy peptides to search for every target peptide
searched. Default 1. Multiple decoys are put in the same collection,
sorted together, and written to the same file.  Does not affect
--number-decoy-sets. 

<li>
<code>mod &lt;mass change&gt;:&lt;aa list&gt;:&lt;max per peptide&gt;
   &ndash; </code>
Consider modifications on any amino acid in aa list with at most
max-per-peptide in one peptide.  This parameter may be included
with different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11.  The same modifications must be given for any post-search
process (compute-q-values, q-ranker).  

<li>
<code>cmod &lt;mass change&gt;:&lt;max distance from protein C-terminus&gt;
&ndash;</code> 
Consider modifications on the C-terminus of any peptide whose
C-terminus is no more than max-distance residues from the protein
C-terminus.  Use -1 to consider the C-terminus all peptides regardless
of position in the protein.  This parameter may be included with
different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11. The same modifications must be given for any
post-search process (compute-q-values, q-ranker).

<li>
<code>nmod &lt;mass change&gt;:&lt;max distance from protein N-terminus&gt;
&ndash;</code> 
Consider modifications on the N-terminus of any peptide whose
N-terminus is no more than max-distance residues from the protein
N-terminus.  Use -1 to consider the N-terminus all peptides regardless
of position in the protein. This parameter may be included with
different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11. The same modifications must be given for any
post-search process (compute-q-values, q-ranker).

<li>
<code>max-mods &lt;n&gt; &ndash;</code>
The maximum number of modifications that can be applied to a single
peptide. Default = no limit.

<li>
<code>max-aas-modified &lt;n&gt; &ndash;</code>
The maximum number of modified amino acids that can appear in one
peptide.  Each aa can be modified multiple times. Default = no limit.

<p>
<i>NOTE: the following parameters are also used when creating an index
and must be compatible with any index used.</i></p>

<LI>
<code>min-mass &lt;float&gt;</code> &ndash; 
The minimum neutral mass of the peptides to place in the
index. Default = 200.

<LI>
<code>max-mass &lt;float&gt;</code> &ndash; 
The maximum neutral mass of the peptides to place in
index. Default = 7200.
	
<LI>
<code>min-length &lt;int&gt;</code> &ndash; 
The minimum length of the peptides to place in the index. Default = 6.
	
<LI>
<code>max-length &lt;int&gt;</code> &ndash; 
The maximum length of the peptides to place in the index. Default = 50.

<LI>
<code>--enzyme &lt;trypsin|chymotrypsin|elastase|no-enzyme&gt; </code> &ndash; 
Enzyme to use for <i>in silico</i> digestion of protein sequences
(trypsin, chymotrypsin, elastase, no-enzyme).  Default trypsin. Used
in conjunction with the options digestion and missed-cleavages. Use
'no-enzyme' for non-specific digestion.  

<LI>
<code>--digestion &lt;full-digest|partial-digest&gt;
Degree of digestion used to generate peptides (full-digest,
partial-digest). Either both ends or one end of a peptide must
conform to enzyme specificity rules. Default full-digest., Used in
conjunction with enzyme option when enzyme is not set to to
'no-enzyme'.  

<LI>
<code>missed-cleavages &lt;T|F&gt;</code> &ndash; 
Allow missed cleavage sites within a peptide.  When used with
enzyme is specified; includes peptides containing one or more
potential cleavage sites. Default = F. 

<li>
<code>unique-peptides &lt;T|F&gt;</code> &ndash; 
For peptides appearing in multiple proteins, store a reference to only
one of those proteins.  Default = F.

<LI>
<code>isotopic-mass &lt;average|mono&gt;</code> &ndash; 
Specify the type of isotopic masses to use when calculating the
peptide mass. Default = average.
   
</ul>

<HR></hr>

<P>
<A HREF="crux.html">Crux home</A></P>

</blockquote>
</BODY>
</HTML>
