<html>
<head>
<title>crux search-for-matches</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<H1 ALIGN=CENTER>crux search-for-matches</H1>

<B>Usage:</B>

<blockquote>
<code>crux search-for-matches [options]
&lt;ms2&nbsp;input&nbsp;filename&gt; &lt;protein&nbsp;input&gt;
</code>
</blockquote>

<B>Description:</B>

<blockquote>
<p>
This command searches a protein database with a set of spectra.  For
each spectrum, the precursor mass is computed from the measured
precursor m/z and an assumed charge.  Candidate peptides whose mass
lies within a specified range of the precursor mass are identified.  These
candidate peptides are first ranked with the SEQUEST Sp score.  The
top 500 matches are then re-scored using XCorr.  The input protein
database may either be in FASTA format or it may be a binary index
created by <code>crux create-index</code>.</p>

<p>
<b>Modifications:</b> Crux handles two types of modifications: static
and variable.  Static modifications are a change of mass applied to
a given amino acid in every peptide in which it occurs.  By default, a
static modification of +57 da to cystine (C) is applied.
Variable modifications allow peptides to be generated with and without
a mass change to a given amino acid.  Crux handles variable 
modifications as follows.   
The user specifies an allowed set of amino acid modifications, using
the options <code>mod</code>, <code>cmod</code> and <code>nmod</code>,
which are described below.  Before any search is performed, Crux
generates an exhaustive list of all possible combinations of amino
acid modifications that could be applied to a peptide.  For example,
if the user specifies a modification of +79 which can be present twice
on any peptide and a modification of +30 which can be present only
once, then the list will contain modifications corresponding to +79
on one residue, +79 on two different residues, +30 on one residue, +79
on one and +30 on one, +79 on two and +30 on one.  Subsequently, for
each spectrum, Crux performs one search for each possible combination of
modifications.  For example, if the precursor m/z for a spectrum is
800 da, the charge state is +2, and Crux is considering a modification
of +79, then Crux will retrieve from the database all candidate
peptides whose total mass is close to 221 da.  These candidates are
scored as usual, with a preliminary score and a final score, and the
top <i>n</i> candidate peptides are added to a composite, sorted list
of peptides.  Finally, after all modifications have been searched,
Crux reports for the current spectrum the top <i>m</i> peptides from
the composite list.</p>

</blockquote>

<p>
<B>Input:</B>

<UL>

<LI>
&lt;ms2 &gt; &ndash; The name of the file (in MS2 format) from which
  to parse the spectra. </li>

<LI>
&lt;protein input&gt; &ndash; 
  The name of the file in fasta format or the directory containing a
  protein index from which to retrieve proteins and peptides.</li>

</UL>

<B>Output:</B>

<p>
The program writes files to the folder <code>crux-output</code> by default.
The name of the output folder can be set by the user using the <code>--output-dir</code> option.
The following files will be created:</p>
<ul>

<li>
<b>search.target.csm</b>: a binary file that contains the PSMs and additional
features.  This file can be used as input
to <code>crux&nbsp;compute-q-values</code>
or <code>crux&nbsp;percolator</code>.

<li>
<b>search.target.sqt</b>: an <a href="sqt-format.html">SQT</a> file containing the PSMs

<li>
<b>search.target.txt</b>: a tab-delimited text file containing the PSMs.
See <a href="txt-format.html">txt file format</a> for a list of the fields.
</li>


<li>
<b>log.txt</b>: a log file containing a copy of all messages that were
printed to stderr.

</ul>

<p>
If decoys are enabled using <code>--number-decoy-set</code>, then
three files called binary.csm, binary.sqt and binary.txt are also
produced.</p>


<B>Options:</B>
<UL>

<code>--fileroot &lt;string&gt;</code> &ndash; 
The <code>fileroot</code> string will be added as a prefix to all
output file names. Default = none.</li>

<li>
<code>--output-dir &lt;filename&gt;</code> &ndash; 
The name of the directory where output files will be created.
Default = crux-out.</li>


<li>
<code>--overwrite &lt;T|F&gt;</code> Replace existing files if true (T) or 
fail when trying to overwrite a file if false (F). Default =  F.</li>

<LI>
<code>--number-decoy-set &lt;0|1|2&gt;</code> &ndash; Specify the number
of decoy sets to generate.  Each decoy set of psms will be printed to
a separate file and not included with the target psms.  To return
decoys in the same file as targets, use <code>tdc=T</code> in the
parameter file and control the number of decoys
with <code>num-decoys-per-target</code>. At least one decoy set (in
its own file) is required to run the algorithm 'percolator' in a subsiquent
crux run.  Default = 2. </li>

<li>
<code>--compute-p-values &lt;T | F&gt; &ndash</code>
Estimate the paramters of the score distribution for each spectrum and
compute a p-value for each PSM.  This option must be used in order to
run <code>crux compute-q-values</code> on the results.</li>

<LI>
<code>--spectrum-min-mass &lt;float&gt;</code> &ndash; The lowest spectrum 
m/z to search in the ms2 file. Default = 0.0

<LI>
<code>--spectrum-max-mass &lt;float&gt;</code> &ndash; The highest spectrum 
m/z to search in the ms2 file. Default = no maximum.

<LI> 
<code>--spectrum-charge &lt;1|2|3|all&gt;</code> &ndash; The spectrum 
charges to search. With 'all' every spectrum will be searched and
spectra with multiple charge states will be searched once at each
charge state.  With 1, 2, or 3 only spectra with that charge will be
searched. Default = all. 

<LI><code>--parameter-file &lt;filename&gt;</code> &ndash; A file containing
command-line or additional parameters.  See the <a
href="parameter-file.html">parameter documentation</a> page for
details.

<LI>
<code>--write-parameter-file &lt;T|F&gt;</code> &ndash; Create a file in the
output directory named <code>params.txt</code>.  This file contains
the name and value of all parameters/options for the current
operation.  Not all parameters in the file may have been used in the
operation.  The resulting file can be used with the <code>--parameter-file</code>
option for other crux programs. Default = <code>F</code>.

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; Specify the verbosity of the
current processes. Each level prints the following messages, including
all those at lower verbosity levels: 0-fatal errors, 10-non-fatal
errors, 20-warnings, 30-information on the progress of execution,
40-more progress information, 50-debug info, 60-detailed debug info.
Default = 30.

<li>
<code>--version T</code> &ndash;
Print the version number and quit.  Please note that you must include
the 'T' after --version.

</UL>
<B>Parameter file options:</B>
<ul>
<li>
<code>fragment-mass &lt;average|mono&gt;</code> &ndash; Which isotopes to use
in calcuating fragment ion mass (average, mono). Default = mono.

<li>
<code>mass-window</code> &ndash;
Tolerance used for matching peptides to spectra.  Peptides must be
within +/- 'mass-window' of the spectrum mass.  Default = 3.0.

<li>
<code>ion-tolerance&lt;float&gt;</code> &ndash; Tolerance used for matching
observed peaks to predicted fragment ions.  Default = 0.5.

<li>
<code>max-rank-preliminary &lt;int&gt;</code> &ndash;  Number of psms per
spectrum to score after preliminary scoring.  Default = 500.

<li>
<code>top-match &lt;int&gt;</code> &ndash; The number of psms per
spectrum writen to the output files.
Default = 5.
<p>
<i>NOTE: <code>crux percolator</code> requires that <code>search-for-matches</code>
be run with <code>top-match=1</code>.</i></p>

<li>
<code>tdc &lt;T|F&gt;</code> &ndash;
Target-decoy competition.  Combine target and decoy searches in one
file.  Default FALSE, separate files. After searching target and decoy
peptides, all psms for one spectrum are sorted together and the mixed
list is returned in the search.target.[sqt|txt|csm] file. Set number of decoys with
num-decoys-per-target.  The number of psms retained after preliminary
searching (max-rank-preliminary) is increased to reflect the number of
decoys searched so that (1+num-decoys)*max-rank-preliminary are kept.
Default = F.

<Li>
<code>num-decoys-per-target &lt;int&gt;</code> &ndash;
Number of decoy peptides to search for every target peptide
searched. Multiple decoys are put in the same collection,
sorted together, and written to the same file.  Does not affect
--number-decoy-sets or tdc.  Default 1. 

<li>
<code>mod &lt;mass change&gt;:&lt;aa list&gt;:&lt;max per peptide&gt;
   &ndash; </code>
Consider modifications on any amino acid in aa list with at most
max-per-peptide in one peptide.  This parameter may be included
with different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11.  The same modifications must be given for any post-search
process (compute-q-values, q-ranker).  

<li>
<code>cmod &lt;mass change&gt;:&lt;max distance from protein C-terminus&gt;
&ndash;</code> 
Consider modifications on the C-terminus of any peptide whose
C-terminus is no more than max-distance residues from the protein
C-terminus.  Use -1 to consider the C-terminus all peptides regardless
of position in the protein.  This parameter may be included with
different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11. The same modifications must be given for any
post-search process (compute-q-values, q-ranker).

<li>
<code>nmod &lt;mass change&gt;:&lt;max distance from protein N-terminus&gt;
&ndash;</code> 
Consider modifications on the N-terminus of any peptide whose
N-terminus is no more than max-distance residues from the protein
N-terminus.  Use -1 to consider the N-terminus all peptides regardless
of position in the protein. This parameter may be included with
different values multiple times so long as the total number of
<code>mod</code>, <code>cmod</code>, and <code>nmod</code> parameters
does not exceed 11. The same modifications must be given for any
post-search process (compute-q-values, q-ranker).

<li>
<code>max-mods &lt;n&gt; &ndash;</code>
The maximum number of modifications that can be applied to a single
peptide. Default = no limit.

<li>
<code>max-aas-modified &lt;n&gt; &ndash;</code>
The maximum number of modified amino acids that can appear in one
peptide.  Each aa can be modified multiple times. Default = no limit.

<li>
<code>precision &lt;n&gt; &ndash; </code>
Set the precision for masses and scores written to sqt and text
files. Default 6. Available from parameter file for crux
search-for-matches, percolator, and compute-q-values. 

<li>
<code>print-search-progress &lt;n&gt; &ndash; </code>
Show search progress by printing every n spectra searched. Set to
0 to show no search progress.  Available for crux search-for-matches
from parameter file. Default = 10.

<li>
<code>reverse-sequence &lt;T|F&gt; &ndash;</code>
Generate decoy sequences by reversing the peptide rather than by
shuffling.  The first and last residues of the sequence are not
changed.  If the target sequence is a palindrome (the same when
reversed), then the decoy will be generated by shuffling and a note to
that effect will be printed at verbosity level 40 (DETAILED INFO).

<p>
<i>NOTE: the following parameters are also used when creating an index
and must be compatible with any index used.</i></p>

<LI>
<code>min-mass &lt;float&gt;</code> &ndash; 
The minimum neutral mass of the peptides to place in the
index. Default = 200.

<LI>
<code>max-mass &lt;float&gt;</code> &ndash; 
The maximum neutral mass of the peptides to place in
index. Default = 7200.
	
<LI>
<code>min-length &lt;int&gt;</code> &ndash; 
The minimum length of the peptides to place in the index. Default = 6.
	
<LI>
<code>max-length &lt;int&gt;</code> &ndash; 
The maximum length of the peptides to place in the index. Default = 50.

<LI>
<code>enzyme &lt;trypsin|chymotrypsin|elastase|clostripain|cyanogen-bromide|idosobenzoate|proline-endopeptidase|staph-protease|modified-chymotrypsin|elastase-trypsin-chymotrypsin|no-enzyme&gt; 
</code> &ndash; 
Enzyme to use for <i>in silico</i> digestion of protein sequences. Used
in conjunction with the options digestion and missed-cleavages. Use
'no-enzyme' for non-specific digestion.  Digestion rules are as
follows: enzyme name [cuts after one of these residues]|{but not before
one of these residues}.  trypsin [RK]|{P}, elastase [ALIV]|{P},
chymotrypsin [FWY]|{P}, clostripain [R]|[], cyanogen-bromide [M]|[],
iodosobenzoate [W]|[], proline-endopeptidase [P]|[], staph-protease
[E]|[], modified-chymotrypsin [FWYL]|{P},
elastase-trypsin-chymotrypsin [ALIVKRWFY]|{P},aspn []|[D] (cuts before
D). Default = trypsin.

<li>
<code>custom-enzyme &lt;residues before cleavage | residues after cleavage
&gt;</code> &ndash
Specify rules for <i>in silico</i> digestion of protein sequences.
Overrides the <code>enzyme</code> option.  Two lists of residues are
given enclosed in square brackets or curly braces and separated by a
|. The first list contains residues required/prohibited before the
cleavage site and the second list is residues after the cleavage site.  
If the residues are required for digestion, they are in square
brackets, '[' and ']'.  If the residues prevent digestion, then they
are enclosed in curly braces, '{' and '}'.  Use X to indicate all
residues.  For example, trypsin cuts after R or K but not before P
which is represented as <code>[RK]|{P}</code>.  AspN cuts after any
residue but only before D which is represented as <code>[X]|[D]</code>.

<LI>
<code>digestion &lt;full-digest|partial-digest&gt;</code> &ndash;
Degree of digestion used to generate peptides (full-digest,
partial-digest). Either both ends or one end of a peptide must
conform to enzyme specificity rules. Default full-digest., Used in
conjunction with enzyme option when enzyme is not set to to
'no-enzyme'.  

<LI>
<code>missed-cleavages &lt;T|F&gt;</code> &ndash; 
Allow missed cleavage sites within a peptide.  When used with
enzyme is specified; includes peptides containing one or more
potential cleavage sites. Default = F. 

<li>
<code>unique-peptides &lt;T|F&gt;</code> &ndash; 
For peptides appearing in multiple proteins, store a reference to only
one of those proteins.  Default = F.

<LI>
<code>isotopic-mass &lt;average|mono&gt;</code> &ndash; 
Specify the type of isotopic masses to use when calculating the
peptide mass. Default = average.
   
<li>
<code>&lt;A-Z&gt &lt;float&gt;</code> &ndash;
Specify static modifications.  This is a mass change applied to the
given amino acid (in single-letter-code A thru Z) for every peptide in
which it occurs.  Use the <code>mod</code> option for generating
peptides both with and without the mass change.  Default C=57.
</ul>

<HR></hr>

<P>
<A HREF="crux.html">Crux home</A></P>

</blockquote>
</BODY>
</HTML>
