<html>
<head>
<title>crux search-for-xlinks</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<H1 ALIGN=CENTER>crux search-for-xlinks</H1>
<!-- open source warning
<h2>
You are using the open source version of Crux.
Due to intellecutal property issues,
we are unable to provide database search functionality in this version.
To obtain a license for the fully functional version of Crux that includes
the database search tools please visit 
<a href="http://depts.washington.edu/ventures/UW_Technology/Express_Licenses/crux.php">UW TECHTRANSFER</a>.
</h2>
open source warning -->
<B>Usage:</B>

<blockquote>
<code>crux search-for-xlinks [options]
&lt;ms2&nbsp;input&nbsp;filename&gt; &lt;protein&nbsp;input&gt;&nbsp;&lt;link&nbsp;sites&gt;&nbsp;&lt;link mass&gt;
</code>
</blockquote>

<B>Description:</B>

<blockquote>
<p>
This command searches a protein database with a set of spectra.  For
each spectrum, the precursor mass is computed from either the precursor 
singly charged mass (m+h) or the mass-to-charge (m/z) and an assumed charge.  
Candidates molecules are linear peptides, dead-end products, self-loop 
products or cross-linked products whose mass lies within a specified range of the precursor
mass.  These candidate peptides are ranked using XCorr.  The input
protein database is in FASTA format.</p>

<p>
The algorithm is described in more detail in the following article:</p>

<blockquote>
Sean McIlwain, Paul Draghicescu, Pragya Singh, David R. Goodlett and
William Stafford
Noble.  <a href="http://noble.gs.washington.edu/papers/mcilwain2010detecting.html">"Detecting
cross-linked peptides by searching against a database of cross-linked
peptide pairs."</a> <i>Journal of Proteome Research</i>.  2010.
</blockquote>

<p>
<b>Modifications:</b> Currently, <code>crux search-for-xlinks</code>
supports static modifications (a change of mass applied to a given
amino acid in every peptide in which it occurs) but not variable
modifications (allowing peptides to be generated with and without a
mass change to a given amino acid).  By default, a static modification
of +57 da to cystine (C) is applied.  Static modifications can be
specified in the parameter file, as described below.</p>
</blockquote>

<p>
<B>Input:</B>

<UL>

<LI>
&lt;ms2 &gt; &ndash; The name of the file from which
  to parse the spectra. </li>

<LI>
&lt;protein input&gt; &ndash; 
  The name of the file in fasta format from which to retrieve proteins and peptides.
</li>

<LI>
&lt;link sites&gt; &ndash; 
A comma delimited list of the amino acid to allow cross-links
with. For example A:K,A:D which means that the cross linker can attach A to
K or A to D.  Also, the n-terminus of a protein can be specified as a link
site by using nterm.  For example  nterm:K means that a cross-linker can
attach between a protein's n-terminus and lysine.

<LI>
&lt;link mass&gt; &ndash; 
  The mass modification of the linker when attached to a peptide.
</UL>

<B>Output:</B>

<p>
The program writes files to the folder <code>crux-output</code> by default.
The name of the output folder can be set by the user using the <code>--output-dir</code> option.
The following files will be created:</p>
<ul>

<li>
<b>search-for-xlinks.params.txt</b>:
a file containing the name and value of all parameters/options for the current operation.  
Not all parameters in the file may have been used in the operation.  
The resulting file can be used with the <code>--parameter-file</code>
option for other crux programs.
</li>

<li>
<b>search.target.txt</b>: a tab-delimited text file containing the PSMs.
See <a href="xlink-txt-format.html">xlink txt file format</a> for a list of the fields.
</li>

<li>
<b>search.decoy.txt</b>: a tab-delimited text file containing the decoy PSMs.
See <a href="xlink-txt-format.html">xlink txt file format</a> for a list of the fields.
</li>

<li>
<b>qvalues.target.txt</b>: a tab-delimited text file containing the top ranked PSMs with calculated q-values.
See <a href="xlink-txt-format.html">xlink txt file format</a> for a list of the fields.
</li>

<li>
<b>search-for-xlinks.log.txt</b>: a log file containing a copy of all messages that were
printed to stderr.

</ul>

<B>Options:</B>
<UL>

<li>
<code>--output-dir &lt;filename&gt;</code> &ndash; 
The name of the directory where output files will be created.
Default = crux-output.</li>


<li>
<code>--overwrite &lt;T|F&gt;</code> Replace existing files if true (T) or 
fail when trying to overwrite a file if false (F). Default =  F.
</li>

<LI>
<code>--spectrum-min-mass &lt;float&gt;</code> &ndash; The lowest spectrum 
m/z to search in the ms2 file. Default = 0.0.
</LI>

<LI>
<code>--spectrum-max-mass &lt;float&gt;</code> &ndash; The highest spectrum 
m/z to search in the ms2 file. Default = no maximum.
</LI>

<LI> 
<code>--spectrum-charge &lt;1|2|3|all&gt;</code> &ndash; The spectrum 
charges to search. With 'all' every spectrum will be searched and
spectra with multiple charge states will be searched once at each
charge state.  With 1, 2, or 3 only spectra with that charge will be
searched. Default = all. 
</LI>

<li>
<code>--compute-sp &lt;T|F&gt;</code> &ndash;
Have <code>search-for-matches</code> compute the Sp score for all
candidate peptides, which is recommended if results are to be analyzed
by <code>percolator</code> or <code>q-ranker</code>.  
<code>sequst-search</code> always computes Sp.  Default = F.

<LI><code>--parameter-file &lt;filename&gt;</code> &ndash; A file containing
command-line or additional parameters.  See the <a
href="parameter-file.html">parameter documentation</a> page for
details.
</LI>

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; Specify the verbosity of the
current processes. Each level prints the following messages, including
all those at lower verbosity levels: 0-fatal errors, 10-non-fatal
errors, 20-warnings, 30-information on the progress of execution,
40-more progress information, 50-debug info, 60-detailed debug info.
Default = 30.
</LI>

<li>
<code>--precursor-window</code> &ndash;
Tolerance used for matching peptides to spectra.  Peptides must be
within +/- '<code>precursor-window</code>' of the spectrum mass.  Definition of
precursor window depends upon <code>precursor-window-type</code>. Default = 3.0.
</li>

<li>
<code>--precursor-window-type &lt;mass|mz|ppm&gt;</code> &ndash; Specify
the units for the window that is used to select peptides around the
precursor mass location (mass, mz, ppm).  The magnitude of the window
is defined by the <code>precursor-window</code> option, and candidate peptides must
fall within this window.  For the <code>mass</code> window-type, the
spectrum precursor m+h value is converted to mass, and the window is defined
as that mass &plusmn; <code>precursor-window</code>. If the m+h value is not available, then
the mass is calculated from the precursor m/z and provided charge. The peptide mass is computed as the
sum of the average amino acid masses plus 18 Da for the terminal OH
group.  The <code>mz</code> window-type calculates
the window as spectrum precursor m/z &plusmn; <code>precursor-window</code> and then 
converts the resulting m/z range to the peptide mass range using the precursor 
charge. 
For the parts-per-million (<code>ppm</code>) window-type, the spectrum mass is 
calculated as in the <code>mass</code> type.  The lower bound of the mass 
window is then defined as the 
spectrum mass / (1.0 + (<code>precursor-window</code> / 1000000)) and the upper bound 
is defined as spectrum mass / (1.0 - (<code>precursor-window</code> / 1000000)).  
Default = mass.</li>

<li><code>--precursor-window-decoy &lt;0-1e6&gt;</code> &ndash; Search
decoy peptides within +/- '<code>precursor-window-decoy</code>' of the precursor mass.
Default = 20.
</li>

<li><code>--precursor-window-type-decoy &lt;mass, mz, ppm&gt;</code> &ndash;
Precursor window type for selecting decoy peptides from precursor mz.
Default = mass.
</li>

<li><code>--min-weibull-points &lt;int&gt;</code> &ndash;
Keep reshuffling and collecting XCorr scores until the minimum number 
of points for weibull fitting (using targets and decoys) is achieved.
Default = 4000.
</li>

<li><code>--max-ion-charge &lt;int&gt;</code> &ndash;
Predict ions for the theoretical spectra up to max charge state 
(1,2,...,6) or up to the charge state of the peptide (peptide).  
If the max-ion-charge is greater than the charge state of the 
peptide, then the max is the peptide charge.  Default='peptide'.

<li>
<code>mod-mass-format &lt;mod-only|total|separate&gt;<code> &ndash;
Specify how sequence modifications are reported in various ouptut
files.  Each modification is reported as a number enclosed in square
braces following the modified reside; however, the number may
correspond to one of three different masses: (1) 'mod-only' reports
the value of the mass shift induced by the modification; (2) 'total'
reports the mass of the residue with the modification (residue mass
plus modification mass); (3) 'separate' is the same as 'mod-only', but
multiple modifications to a single amino acid are reported as a
comma-separated list of values. For example, suppose amino
acid <i>D</i> has an unmodified mass of 115 as well as two
modifications of masses +14 and +2.  In this case, the amino acid
would be reported as D[16] with 'mod-only', D[131] with 'total', and
D[14,2] with 'separate'.</li>

<!--
<li><code>--missed-link-cleavage &lt;string&gt;</code> &ndash;
Currently unused parameter. Default = K.
</li>
-->

<li><code>--xlink-include-linears &lt;T|F&gt;</code> &ndash;
Include linear peptides in the search. Default = T.
</li>

<li><code>--xlink-include-deadends &lt;T|F&gt;</code> &ndash;
Include deadend products in the search. Default = T.
</li>

<li><code>--xlink-include-selfloops &lt;T|F&gt;</code> &ndash;
Include selfloops in the search. Default = T.
</li>

<li>
<code>--xlink-prevents-cleavage &lt;string&gt;</code> &ndash;
List of amino acids that xlinker can prevent cleavage.

<li>
<code>--use-flanking-peaks &lt;T|F&gt;</code> &ndash;
Turn on or off the peaks flanking the b/y ions.  For <code>crux 
search-for-matches</code>, default = F; for <code>crux
  sequest-search</code>, default = T; for <code> crux
search-for-xlinks</code>, default = T.</li>

<li><code>--xcorr-use-flanks &lt;T|F&gt;</code> &ndash;
Use flanking ions in the theoretical spectrum.  These are placed +/- 1
  Da around the b-y ions, with intensity of 25.0.
Default = T.
</li>

<li><code>--use-mgf &lt;T|F&gt;</code> &ndash;
Use MGF file format for parsing spectra. Default = F.
</li>

<li>
<code>--top-match &lt;int&gt;</code> &ndash; The number of PSMs per
spectrum written to the output files. Default = 5.
</li>

</UL>
<B>Parameter file options:</B>
<ul>
<li>
<code>fragment-mass &lt;average|mono&gt;</code> &ndash; Which isotopes to use
in calcuating fragment ion mass (average, mono). Default = average.


<li>
<code>use-mstoolkit</code> &ndash;
Use MSToolkit to parse the spectra. Default = F.
</li>

<LI>
<code>min-mass &lt;float&gt;</code> &ndash; 
The minimum neutral mass of the peptides to place in the
index. Default = 200.

<LI>
<code>max-mass &lt;float&gt;</code> &ndash; 
The maximum neutral mass of the peptides to place in
index. Default = 7200.
	
<LI>
<code>min-length &lt;int&gt;</code> &ndash; 
The minimum length of the peptides to place in the index. Default = 4.
	
<LI>
<code>max-length &lt;int&gt;</code> &ndash; 
The maximum length of the peptides to place in the index. Default = 50.

<LI>
<code>enzyme &lt;trypsin|chymotrypsin|elastase|clostripain|cyanogen-bromide|idosobenzoate|proline-endopeptidase|staph-protease|modified-chymotrypsin|elastase-trypsin-chymotrypsin|no-enzyme&gt; 
</code> &ndash; 
Enzyme to use for <i>in silico</i> digestion of protein sequences. Used
in conjunction with the options digestion and missed-cleavages. Use
'no-enzyme' for non-specific digestion.  Digestion rules are as
follows: enzyme name [cuts after one of these residues]|{but not before
one of these residues}.  trypsin [RK]|{P}, elastase [ALIV]|{P},
chymotrypsin [FWY]|{P}, clostripain [R]|[], cyanogen-bromide [M]|[],
iodosobenzoate [W]|[], proline-endopeptidase [P]|[], staph-protease
[E]|[], modified-chymotrypsin [FWYL]|{P},
elastase-trypsin-chymotrypsin [ALIVKRWFY]|{P},aspn []|[D] (cuts before
D). Default = trypsin.

<li>
<code>custom-enzyme &lt;residues before cleavage | residues after cleavage
&gt;</code> &ndash
Specify rules for <i>in silico</i> digestion of protein sequences.
Overrides the <code>enzyme</code> option.  Two lists of residues are
given enclosed in square brackets or curly braces and separated by a
|. The first list contains residues required/prohibited before the
cleavage site and the second list is residues after the cleavage site.  
If the residues are required for digestion, they are in square
brackets, '[' and ']'.  If the residues prevent digestion, then they
are enclosed in curly braces, '{' and '}'.  Use X to indicate all
residues.  For example, trypsin cuts after R or K but not before P
which is represented as <code>[RK]|{P}</code>.  AspN cuts after any
residue but only before D which is represented as <code>[X]|[D]</code>.

<LI>
<code>digestion &lt;full-digest|partial-digest&gt;</code> &ndash;
Degree of digestion used to generate peptides (full-digest,
partial-digest). Either both ends or one end of a peptide must
conform to enzyme specificity rules. Default full-digest., Used in
conjunction with enzyme option when enzyme is not set to to
'no-enzyme'.  

<LI>
<code>missed-cleavages &lt;T|F&gt;</code> &ndash; 
Allow missed cleavage sites within a peptide.  When used with
enzyme is specified; includes peptides containing one or more
potential cleavage sites. Default = F. 
</LI>

<li>
<code>unique-peptides &lt;T|F&gt;</code> &ndash; 
For peptides appearing in multiple proteins, store a reference to only
one of those proteins.  Default = F.

<LI>
<code>isotopic-mass &lt;average|mono&gt;</code> &ndash; 
Specify the type of isotopic masses to use when calculating the
peptide mass. Default = mono.
   
<li>
<code>&lt;A-Z&gt &lt;float&gt;</code> &ndash;
Specify static modifications.  This is a mass change applied to the
given amino acid (in single-letter-code A thru Z) for every peptide in
which it occurs.  Use the <code>mod</code> option for generating
peptides both with and without the mass change.  Default C=57.0214637206.
</ul>


<B>Frequently asked questions:</B>

<blockquote>
<b>What could cause a high q-value even when the PSM gets a high
XCorr?</b><br>

<p>
For some spectra, even randomly generated peptides can get relatively
high XCorr scores.  This is what the empirical curve fitting is
supposed to control for.  The idea is that a high XCorr for one
spectrum means something different than the same XCorr for another
spectrum. Using the curve fit and calculating a p-value for the top
ranked PSM helps to control for the spectrum-specific effects.  As
long the assumption that the distribution of XCorr scores for a
particular spectrum follows a Weibull distribution, then a small
q-value will be statistically significant.</p>

<p>
<b>What happens if you decrease the size of the precursor window?</b></p>

<p>
Reducing the precursor mass tolerance has two main benefits.  One is
reduced search time, and the other is improved statistical power to
detect matches.  With a smaller window, fewer candidates are tested
against a spectrum, which in turn means fewer tests.  As a result, the
statistical confidence measure calculated after multiple testing
correction will be more significant.  Of course, the flipside is that
if you make the precursor window too small, then you may end up
throwing out correct identifications.  Control over the size of the
window is provided by the using the <code>precursor-window</code> and
<code>precursor-window-type</code> options.</p>

<p>
In order for the p-value to be accurately estimated, you may need many
data points for fitting the Weibull distribution.  To achieve an
accurate estimate, this program generates decoy peptides until it
obtains <code>min-weibull-points</code> XCorr scores (default 4000).
The decoys are generated by shuffling target peptides from the
database using a larger precursor window. The window size for the
decoys is obtained from the <code>precursor-window-decoy</code>
and <code>precursor-window-type-decoy</code> parameters.</p>

</blockquote>

<hr></hr>

<P>
<A HREF="crux.html">Crux home</A></P>

</blockquote>
</BODY>
</HTML>
