<!--
FILE: customize-tutorial.html 
AUTHOR: Barbara Frewen
DATE: Octobeer 15, 2007
PROJECT: crux
REVISION: $1.0 $
-->

<html>
<head>
<title>Crux: Customization</title>

<link rel="stylesheet" href="stylesheet.css" type="text/css">

</head>

<body>
<div id="top">
<h>crux: peptide spectrum analysis tools</h>
</div>

<div id="main">
<div id="side">
<h3>Table of Contents</h3>

<dl>
   <dt><a href="home-tutorial.html">Home</a>
   <dt><a href="install-tutorial.html">Installation</a>
   <dt><a href="start-tutorial.html">Getting Started</a>
   <dt><a href="search-tutorial.html">Searching a Protein Database</a>
   <dt><a href="index-tutorial.html">Creating an Index</a>
   <dt>Customizing a Search
<!--
   <dt><a href="utilities-tutorial.html">Crux utilities</a>
   <dt><a href="examples-tutorial.html">Examples</a>
-->
   <dt><a href="crux.html">Man pages</a>
<!--   <dt><a href="sitemap-tutorial.html">Index</a> -->
</dl>
<hr>
<h3>On this Page</h3>
<dl>
   <dt><a href="#intro">Introduction to options</a>
   <dt><a href="#file">Parameter files</a>
   <dt><a href="#out">Controlling the output</a>
   <dt><a href="#pep">Change the peptide parameters</a>
   <dt><a href="#search">Change the search parameters</a>
   <dt><a href="#anal">Change the analysis parameters</a>
</dl>
</div><!--end side-->

<h3>Customization and Search Options</h3>
<p><span class="prog">Crux</span> allows the user the flexibility to
change many of the search and analysis parameters.  Attributes like
the format of output and which peptides are selected from the protein
database are controlled through the numerous options.  This page will
start with some general information about options and then describe
the use of some key <span class="prog">crux</span> options.
</p>

<a name="intro"></a>
<h4>Introduction to options</h4>
<p>A <span class="prog">crux</span> command is made up of three parts:
executable name, required arguments, and options. Let's use a <span
class="prog">crux-search-for-matches</span> command as an example.
Here is the general form.</p>

<div class="code"><code>
$ crux-search-for-matches [options] &lt;ms2 filename&gt; &lt;protein
filename&gt;
</code>
</div>

<p>In this case, the name is <span
class="prog">crux-search-for-matches</span>.  The required arguments
are given inside of angle brackets, &lt;&gt;, and are the name of the
input ms2 file and the name of the input protein fasta file.  Options
are just that, optional instructions.  They are always placed between
the executable name and the arguments.  Any number of options can be
included separated by spaces.

<p>All of the available options are described for each executable on
the <a href="crux.html">documentation</a> pages.  You can also get a
list of available options by running a command with just the name and
no arguments.
</p>

<div class="code"><code>
$ crux-search-for-matches</code>
</div>
<p>You should see output that looks like this.</p>
<pre>
Error in command line. Error # 5
the required argument &lt;ms2 file&gt; is missing.
Usage: crux-search-for-matches [options] &lt;ms2 file&gt; &lt;protein input&gt;
Optional Arguments:
  [--verbosity &lt;int&gt;] Set level of output to stderr (0-100).  Default 30.
  [--parameter-file &lt;string&gt;] Set additional options with values in the given file.
  [--overwrite &lt;string&gt;] Replace existing files (T,F). Default F, die if a file of the same name as the output exists.
  [--use-index &lt;string&gt;] Use an index that has already been created (T,F). Default FALSE (use fasta file)
  [--prelim-score-type &lt;string&gt;] Initial scoring (sp, xcorr). Default sp
  [--score-type &lt;string&gt;] The scoring method to use (xcorr, xcorr_logp, sp_logp). Default xcorr.
  [--spectrum-min-mass &lt;double&gt;] Minimum mass of spectra to be searched.  Default 0.
  [--spectrum-max-mass &lt;double&gt;] Maximum mass of spectra to search.  Default, none.
  [--spectrum-charge &lt;string&gt;] Spectrum charge states to search (1,2,3,all). Default all.
  [--match-output-folder &lt;string&gt;] Folder to which search results will be written.  Default '.' (current dir)
  [--output-mode &lt;string&gt;] Types of output to produce (binary, sqt, all). Default binary
  [--sqt-output-file &lt;string&gt;] SQT output file name. Default 'target.sqt'
  [--decoy-sqt-output-file &lt;string&gt;] SQT output file name for decoys.  Default 'decoy.sqt'
  [--number-decoy-set &lt;int&gt;] The number of decoy databases to search.  Default 2.
Required Arguments:
  &lt;ms2 file&gt; File containing spectra to be searched.
  &lt;protein input&gt; File containing protein sequences either in fasta format or binary index.
</pre>

<p>The first three lines are telling you that you forgot the required
arguments and are reminding you what they are.  The following lines
list all the options.  <span class="prog">Crux</span> options all
begin with two dashes, <span class="out">--</span>, followed by the
name (with no space in between).  The name is followed by a space and the
appropriate argument.  This example increases the verbosity to 40
</p>

<div class="code"><code>
$ crux-search-for-matches --verbosity 40 sample.ms2 yeast.fasta</code>
</div>

<a name="file"></a>
<h4>Parameter files</h4>
<p>The second option listed above, <span
class="out">--parameter-file</span>, is available for all <span
class="prog">crux</span> programs.  It allows options to be specified in
a file.  All of the command line options as well as some extras can be
put in a parameter file.  The format is slightly different.  The two
dashes are removed from the option name and the name and value are
separated by an equal sign instead of a space. 
</p>
<div class=code><code>
&lt;option name&gt;=&lt;option argument (value)&gt;</code>
</div>

<p>
The parameter file also requires only one option per line and allows
comments on lines beginning with '#'.  More information and a sample
parameter file can be found <a href="parameter-file.html">here</a>.
Command line and parameter file options may be used separately or
together.  If an option is specified in both, the value on the command
line will be used.
</p>

<p>Now that we've covered the general form of options, let's discuss
some specific options for <span class="prog">crux-search-for-matches</span>
programs, beginning with output.
</p>

<a name="out"></a>
<h4>Controlling the output</h4>
<p><span class="prog">Crux-search-for-matches</span> prints results in
either a binary or text (sqt) format and is controlled with the option <span
class="code">--output-mode</span> and values <span
class="code">binary, sqt,</span> or <span class="code">all</span> (for
both binary and text outputs).
The option <span class="code">--overwrite</span> decides if output
files can be replaced if they already exist (value T) or if the
program should not overwrite an existing file (value F).  In the later
case, the search will fail if a file of the same name already exists.
The text files can be named with  
<span class="code">--sqt-output-file</span> for the matches to the
target database and <span class="code">--decoy-sqt-output</span>
for matches to the shuffled database.
</p>

<a name="pep"></a>
<h4>Changing the peptide parameters</h4>
<p>Several parameters control which peptides in the database are used
in the search.  By setting, for example, <span
class="code">--min-length 6</span>
and <span class="code">--max-length 50</span> you confine the search
to only those peptides with at least 6 amino acids and no more than
50.  Similarly <span class="code">--min-mass</span> and
<span class="code">--max-mass</span> set limits on peptides based on
their calculated masses.  The masses can be calculated as
mono-isotopic or average with the <span
class="code">--isotopic-mass</span> option and the values <span
class="code">mono</span> or <span class="code">average</span>.</p>

<!--isotopic mass, cleavages, and missed cleavages are not allowed on -->
<!--the command line for search-for-matches.  Do they still get read -->
<!--from the file?  Are they considered if searching a fasta file? -->
<!--Are the settings for an index used and do they override given options? -->

<div class="note"><em>NOTE: Options for enzyme specificity and peptide
length and mass limits must correspond with those used with
crux-create-index, if an index is being used with the search.</em>
</div>
<p>Many protein samples are digested with trypsin and the peptides
searched can reflect this.  Use the option <span class="code">--cleavages
</span> with the value <span class="code">tryptic</span> to search
only peptides with trypsin-specific cleavage sites at the C and N
termini, <span class="code">partial</span> to search peptides with at
least one tryptic site, <span class="code">all</span> to use no enzyme
specificity.  Use the <span class="code">--missed-cleavages T</span> 
option to allow tryptic cleavage sites within the peptide sequence (F
to require complete digestion).
</p>

<a name="search"></a>
<h4>Changing the search parameters</h4>
<p>The database search involves two scoring steps, a preliminary score
to find the 500 best candidates and a main scoring function.  Set the
preliminary scoring function with <span
class="code">--prelim-score-type</span> and the main scoring function
with <span class="code">--score-type</span>.</p>
<p>
In addition to searching the provided protein database, 
<span class="prog">crux-search-for-matches</span> will search up to
3 shuffled versions of the database with the option 
<span class="code">--number-decoy-set</span>.  The default number is 2.
</p>

<a name="anal"></a>
<h4>Changing the analysis parameters</h4>
<p>The key option for <span class="prog">crux-analyze-matches</span>
is <span class="code">--algorithm</span> which controls how the
analysis is performed.  The default value is <span
class="code">percolator</span> which will use the <a
href="http://noble.gs.washington.edu/proj/percolator/">Percolator</a>
algorithm to look at target and decoy
matches and calculate a q-value for each spectrum's best match.
Another possibility, <span class="code">qvalue</span>, is to calculate
a q-value by estimating the parameters of the score distribution (see
<a href="http://www.pnas.org/cgi/content/full/100/16/9440">this
paper</a> as an example).
</p>

<span class="code"></span>







</div><!--end main-->

</body>
</html>
