<html>
<head>
<title>crux generate-decoys</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<h1 align=center>crux generate-decoys</h1>

<p>
<b>Usage:</b></p>

<blockquote>
<code>crux generate-decoys [options]  &lt;protein&nbsp;input&nbsp;file&gt;</code>
</blockquote>

<p>
<b>Description:</b></p>

<blockquote>

<p>
This command takes as input a protein FASTA file and outputs the
corresponding list of peptides, as well as a matched list of decoy
peptides and decoy proteins.  Decoys are generated either by reversing
or shuffling the non-terminal amino acids of each peptide.  The
program will shuffle each peptide multiple times to attempt to ensure
that there is no overlap between the target and decoy peptides.  For
homopolymers, this is not possible.  In this case, the occurrence of
these target/decoy overlaps is recorded in the log file.</p>
</blockquote>

<p>
<b>Input:</b></p>

<ul>

<li>
&lt;protein&nbsp;input&nbsp;file&gt; &ndash; A file in FASTA format
containing proteins.</li>

</ul>

<b>Output:</b>

<ul>
<p>
The program writes files to the folder <code>crux-output</code> by
default.  The name of the output folder can be set by the user using
the <code>--output-dir</code> option.  The following files will be
created:</p>

<li>
<b>peptides.target.txt</b>: a text file containing the target
peptides, one per line, sorted by mass.</li>

<li>
<b>peptides.decoy.txt</b>: a text file containing the decoy peptides,
one per line, sorted by mass.  There is a one-to-one correspondence
between targets and decoys.</li>

<li>
<b>proteins.decoy.txt</b>: a FASTA format file containing decoy
proteins, in which all of the peptides have been replaced with their
shuffled or reversed counterparts.  Note that this file will only be
created if the enzyme specificity is "full-digest" and no missed
cleavages are allowed.</li>

<li>
<b>generate-decoys.params.txt</b>: a file containing the name and value of all
parameters/options for the current operation.  Not all parameters in
the file may have been used in the operation.  The resulting file can
be used with the <code>--parameter-file</code> option for other crux
programs.</li>

<li>
<b>tide-search.log.txt</b>: a log file containing a copy of all messages
that were printed to the screen during execution.</li>

</ul>

<b>Options:</b>

<ul>

<LI>
<code>--min-mass &lt;float&gt;</code> &ndash;
The minimum neutral mass of the peptides to place in the
index. Default = 200.
	
<LI>
<code>--max-mass &lt;float&gt;</code> &ndash;
The maximum neutral mass of the peptides to place in index. Default =
7200.

<LI>
<code>--min-length &lt;int&gt;</code> &ndash;
The minimum length of the peptides to place in the index. Default = 6.
	
<LI>
<code>--max-length &lt;int&gt;</code> &ndash;
The maximum length of the peptides to place in the index. Default = 50.

<LI>
<code>--enzyme trypsin|chymotrypsin|elastase|clostripain|cyanogen-bromide|idosobenzoate|proline-endopeptidase|staph-protease|modified-chymotrypsin|elastase-trypsin-chymotrypsin|no-enzyme</code> &ndash;
Enzyme to use for <i>in silico</i> digestion of protein
sequences. Used in conjunction with the <code>digestion</code>
and <code>missed-cleavages</code> options. Use 'no-enzyme' for
non-specific digestion.  Digestion rules are as follows: enzyme name
[cuts after one of these residues]|{but not before one of these
residues}.  trypsin [RK]|{P}, elastase [ALIV]|{P}, chymotrypsin
[FWY]|{P}, clostripain [R]|[], cyanogen-bromide [M]|[], iodosobenzoate
[W]|[], proline-endopeptidase [P]|[], staph-protease [E]|[],
modified-chymotrypsin [FWYL]|{P}, elastase-trypsin-chymotrypsin
[ALIVKRWFY]|{P},aspn [X]|[D] (cuts after any residue and before
D). Default = trypsin.

<li>
<code>--custom-enzyme &lt;residues before cleavage&gt;|&lt;residues after cleavage&gt;</code> &ndash;
Specify rules for <i>in silico</i> digestion of protein sequences.
Overrides the <code>enzyme</code> option.  Two lists of residues are
given enclosed in square brackets or curly braces and separated by a
|. The first list contains residues required/prohibited before the
cleavage site and the second list is residues after the cleavage site.
If the residues are required for digestion, they are in square
brackets, '[' and ']'.  If the residues prevent digestion, then they
are enclosed in curly braces, '{' and '}'.  Use X to indicate all
residues.  For example, trypsin cuts after R or K but not before P
which is represented as <code>[RK]|{P}</code>.  AspN cuts after any
residue but only before D which is represented as <code>[X]|[D]</code>.

<LI>
<code>--digestion full-digest|partial-digest </code>&ndash;
Degree of digestion used to generate peptides . Either both ends
(full-digest) or at least one end (partial-digest) of a peptide must
conform to enzyme specificity rules. Used in conjunction with
the <code>enzyme</code> or <code>custom-enzyme</code> option
when <code>enzyme</code> is not set to to 'no-enzyme'.   Default
full-digest.

<LI>
<code>--missed-cleavages &lt;int&gt;</code> &ndash; Include in the
index peptides containing up to &lt;int&gt; missed cleavage sites.
Default = 0.</li>

<LI>
<code>--isotopic-mass average|mono</code> &ndash;
Specify the type of isotopic masses to use when calculating the
peptide mass. Default = average.</li>

<li>
<code>--decoys &lt;string&gt;</code> &ndash; Include a decoy version
of every peptide by shuffling or reversing the target sequence.
Possible values are <code>reverse, shuffle</code>. Use 'reverse' to
reverse each peptide or shuffle to shuffle each peptide.  Note that
the N-terminal and C-terminal amino acids are always left in place.
Default=shuffle.

<li>
<code>--overwrite T|F</code> Replace existing files if true (T) or
fail when trying to overwrite a file if false (F). Default = F.

<LI>
<code>--parameter-file &lt;string&gt;</code>
A file containing command-line or additional parameters.  See the <a
href="parameter-file.html">parameter documentation</a> page for
details.  Default = no parameter file.

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; 
Specify the verbosity of the current processes. Each level prints the
following messages including all those at lower verbosity levels:
0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on
the progress of execution, 40-more progress information, 50-debug
info, 60-detailed debug info. Default = 30.

</ul>
</ul>

<hr></hr>

<p>
<a href="crux.html">Home</a></p>

</blockquote>
</body>
</html>


