<html>
<head>
<title>crux tide-search</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<h1 align=center>crux tide-search</h1>

<p>
<b>Usage:</b></p>

<blockquote>
<code>crux tide-search [options]  &lt;mass&nbsp;spectra&gt; 
  &lt;protein&nbsp;database&gt;</code>
</blockquote>

<p>
<b>Description:</b></p>

<blockquote>

<p>
Tide is a tool for identifying peptides from tandem mass spectra.  It
is an independent reimplementation of the SEQUEST<sup>&reg;</sup>
algorithm, which assigns peptides to spectra by comparing the observed
spectra to a catalog of theoretical spectra derived from a database of
known proteins.  Tide's primary advantage is its speed.
Our published paper provides more detail on how Tide works. If
you use Tide in your research, please cite:

  <blockquote>
  Benjamin J. Diament and William Stafford
  Noble. &ldquo;<a href="http://dx.doi.org/10.1021/pr101196n">Faster
  SEQUEST Searching for Peptide Identification from Tandem Mass
  Spectra.</a>&rdquo;
  <i>Journal of Proteome Research</i>. 10(9):3871-9, 2011.
  </blockquote>

<p>
Note that when Tide runs, it performs several intermediate steps, as
follows:</p>
<ol>
<li>
Convert the given FASTA format protein database to a binary
index.</li>
<li>
Convert the given fragmentation spectra to a binary format.</li>
<li>
Search the spectra against the database and store the results in
binary format.</li>
<li>
Convert the results to one or more requested output formats.</li>
</ol>
By default, the intermediate binary files are stored in a temporary
directory and deleted when Tide finishes execution.  If you plan to
search a given set of spectra more than once or (more commonly) if you
plan to search multiple sets of spectra against the same protein
database, then you can direct Tide to save these intermediate files.
Subsequent runs of the program will go faster if provided with inputs
in binary format.</p>

</blockquote>

<p>
<b>Input:</b></p>

<ul>

<li>
&lt;spectra&gt; &ndash; The name of the file from which to parse the
fragmentation spectra, in any of
the <a href="http://proteowizard.sourceforge.net/tools.shtml">file
formats supported by ProteoWizard</a>.  Alternatively, the argument
may be a binary spectrum file produced by a previous run of <code>crux
tide-search</code> using the <code>store-spectra</code>
parameter.</li>

<li>
&lt;protein&nbsp;database&gt; &ndash; The name of the file in FASTA
format from which to retrieve proteins.  Alternatively, this argument
may be a database directory created by a previous run of <code>crux
tide-search</code> using the <code>store-database</code>
parameter.</li>

</ul>

<b>Output:</b>

<ul>
<p>
The program writes files to the folder <code>crux-output</code> by
default.  The name of the output folder can be set by the user using
the <code>--output-dir</code> option.  The following files will be
created:</p>

<li>
<b>tide-search.target.txt</b>: a tab-delimited text file containing the
PSMs.  See <a href="txt-format.html">txt file format</a> for a list of
the fields.</li>

<li>
<b>tide-search.params.txt</b>: a file containing the name and value of all
parameters/options for the current operation.  Not all parameters in
the file may have been used in the operation.  The resulting file can
be used with the <code>--parameter-file</code> option for other crux
programs.
</li>

<li>
<b>tide-search.log.txt</b>: a log file containing a copy of all messages
that were printed to the screen during execution.</li>

</ul>

<b>Options:</b>

<p><i>Indexing options</i></p>

<ul>

<li>
<code>--enzyme=&lt;string&gt;</code> &ndash; Specify the enzyme used
to digest the proteins <i>in silico</i>.  Available enzymes include
trypsin, chymotrypsin, elastase, clostripain, cyanogen-bromide,
idosobenzoate, proline-endopeptidase, staph-protease,
modified-chymotrypsin, elastase-trypsin-chymotrypsin and aspn.
Specifying <code>--enzyme=none</code> yields a non-enzymatic
digest.  <b>Warning:</b> the resulting index may be quite large.
Default=<code>none</code>.</li>

<li>
<code>--digestion=full-digest|partial-digest</code> &ndash; Specify
whether every peptide in the database must have two enzymatic termini
(<code>full-digest</code>) or if peptides with only one enzymatic
terminus are also included (<code>partial-digest</code>).
Default=<code>full-digest</code></li>

<li>
<code>--max_missed_cleavages</code> &ndash; Maximum number of missed cleavages
per peptide to allow in enzymatic digestion.  If this option is not specified,
then missed cleavages are not allowed.</li>

<li>
<code>--max_length=&lt;integer&gt;</code> &ndash; Specify the maximum
length of the peptides included in the index.  Default=50.</li>

<li>
<code>--max_mass=&lt;integer&gt;</code> &ndash; Specify the maximum
monoisotopic mass (in Da) of the peptides included in the index.
Default=7200.</li>

<li>
<code>--min_length=&lt;integer&gt;</code> &ndash; Specify the minimum
length of the peptides included in the index.  Default=6.</li>

<li>
<code>--min_mass=&lt;integer&gt;</code> &ndash; Specify the minimum
monoisotopic mass (in Da) of the peptides included in the index.
Default=200.</li>

<li>
<code>--monoisotopic_precursor</code> &ndash; Use monoisotopic
precursor masses rather than average mass (default) for
precursor.
</li>

<li>
<code>--mods_spec</code> &ndash; Expression for static and variable
      mass modifications to include. Specify a comma-separated list of
      modification specifications of the form:

      <p><code>--mods_spec=C+57.02146,2M+15.9949,1STY+79.966331</code>,...

      <p>The general form of a modification specification has 3
      components, as exemplified
      by <code><red>1</red><green>STY</green>+<blue>79.966331</blue></code>.

      <p>The three components are:

      <p><code>[<red>max_per_peptide</red>]<green>residues</green>+<blue>added_mass</blue></code>.

      <p>In the example, <code><red>max_per_peptide</red></code>
      is <code><red>1</red></code>, <code><green>residues</green></code>
      are <code><green>STY</green></code>,
      and <code><blue>added_mass</blue></code>
      is <code><blue>79.966331</blue></code>.

<ul>
      <li><code><green>residues</green></code> indicate which amino
      acids may be affected by the modification.

      <li><code><blue>added_mass</blue></code> indicates in Daltons
      the mass to add to modified versions of the residues.

      <li><code><red>max_per_peptide</red></code> indicates how
      many occurrences of the indicated variable modification may
      occur per peptide. The <code><red>max_per_peptide</red></code>
      may be omitted, as in <code>C+57.02146</code>, to indicate a
      static modification, which will be in force at all occurrences
      of the amino acid residue.
</ul>

      <p>Notes: Currently only positive mass modifications are
      specifiable (<code><blue>added_mass</blue></code> &gt; 0). Use
      caution in specifying variable modifications: specifying many
      modifications or higher numbers of max_per_peptide can increase
      index size significantly.

</li>

</ul>

<p><i>Search options</i></p>

<ul>

<li>
<code>--mass_window=&lt;float&gt;</code> &nbsp; Specify the precursor
mass tolerance in Daltons.  Default=3.</li>

<li>
<code>--top_matches=&lt;integer&gt;</code> &nbsp; Specify the number of
matches to report for each spectrum.  Default=5.</li>

</ul>

<p><i>Output options</i></p>

<ul>

<li>
<code>--store-index=&lt;directory&gt;</code> &ndash; Specify the name of
the directory where the protein database index will be stored.  The
directory will contain three files, containing information about
peptides, proteins, and where peptides are located within the
proteins.  By default, these files are stored in a temporary directory
and deleted at the end of the run.</li>

<li>
<code>--store-spectra=&lt;filename&gt;</code> &ndash; Specify the name
of the file where the binarized fragmentation spectra will be stored.
By default, this file is stored in a temporary directory and deleted
at the end of the run.  Subsequent runs of <code>crux
tide-search</code> will execute more quickly if provided with the
spectra in binary format.</li>

<li>
<code>--txt-output &lt;T|F&gt;</code> &ndash; Output a tab-delimited
results file to the output directory. Default = T.</li>

<li>
<code>--sqt-output &lt;T|F&gt;</code> &ndash; Output an SQT
results file to the output directory. Default = F.</li>

<li>
<code>--pepxml-output &lt;T|F&gt;</code> &ndash; Output a pepXML
results file to the output directory. Default = F.</li>

<li>
<code>--fileroot &lt;string&gt;</code> &ndash; 
The <code>fileroot</code> string will be added as a prefix to all
output file names. Default = none.</li>

<li>
<code>--output-dir &lt;filename&gt;</code> &ndash; 
The name of the directory where output files will be created.
Default = crux-out.</li>

<li>
<code>--overwrite T|F</code> &ndash;
Replace existing files if true (T) or fail when trying to overwrite a
file if false (F). Default =  F.</li>

<LI><code>--parameter-file &lt;filename&gt;</code> &ndash; 
A file containing command-line or additional parameters.  See
the <a href="parameter-file.html">parameter documentation</a> page
for details.  Default = no parameter file.</li>

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; 
Specify the verbosity of the current processes. Each level prints the
following messages, including all those at lower verbosity levels:
0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on
the progress of execution, 40-more progress information, 50-debug
info, 60-detailed debug info. Default = 30.</li>

</ul>

<hr></hr>



</blockquote>
</body>
</html>


