<html>
<head>
<title>crux tide-index</title>
<link href="./crux.css" rel="styleSheet" type="text/css">
</head>
<body>
<blockquote>

<h1 align=center>crux tide-index</h1>

<p>
<b>Usage:</b></p>

<blockquote>
<code>crux tide-index [options]
  &lt;protein&nbsp;input&nbsp;file&gt;
  &lt;index&nbsp;name&gt;
</code>
</blockquote>

<p>
<b>Description:</b></p>

<blockquote>

<p>
Tide is a tool for identifying peptides from tandem mass spectra.  It
is an independent reimplementation of the SEQUEST<sup>&reg;</sup>
algorithm, which assigns peptides to spectra by comparing the observed
spectra to a catalog of theoretical spectra derived from a database of
known proteins.  Tide's primary advantage is its speed.
Our published paper provides more detail on how Tide works. If
you use Tide in your research, please cite:

  <blockquote>
  Benjamin J. Diament and William Stafford
  Noble. &ldquo;<a href="http://dx.doi.org/10.1021/pr101196n">Faster
  SEQUEST Searching for Peptide Identification from Tandem Mass
  Spectra.</a>&rdquo;
  <i>Journal of Proteome Research</i>. 10(9):3871-9, 2011.
  </blockquote>

<p>
The <code>tide-index</code> command performs a required pre-processing
step on the protein database, converting it to a binary format
suitable for input to the <code>tide-search</code> command.</p>

</blockquote>

<p>
<b>Input:</b></p>

<ul>

<li>
&lt;protein&nbsp;input&nbsp;file&gt; &ndash; The name of the file in
FASTA format from which to retrieve proteins.</li>

<li>
&lt;index&nbsp;name&gt; &ndash; The desired name of the binary
index.</li>

</ul>

<b>Output:</b>

<ul>

<p>
The program creates a binary index, using the name specified on the
command line.</p>

<p>
In addition, the program writes the following files to the
folder <code>crux-output</code>.  The name of the output folder can be
set by the user using the <code>--output-dir</code> option.  The
following files will be created:</p>

<ul>

<li>
<b>tide-index.params.txt</b>: a file containing the name and value of all
parameters/options for the current operation.  Not all parameters in
the file may have been used in the operation.  The resulting file can
be used with the <code>--parameter-file</code> option for other crux
programs.
</li>

<li>
<b>tide-index.log.txt</b>: a log file containing a copy of all messages
that were printed to the screen during execution.</li>

</ul>

</ul>

<b>Options:</b>

<ul>

<li>
<code>--enzyme &lt;string&gt;</code> &ndash; Specify the enzyme used
to digest the proteins <i>in silico</i>.  Available enzymes include
trypsin, chymotrypsin, elastase, clostripain, cyanogen-bromide,
idosobenzoate, proline-endopeptidase, staph-protease,
modified-chymotrypsin, elastase-trypsin-chymotrypsin and aspn.
Specifying <code>--enzyme no-enzyme</code> yields a non-enzymatic
digest.  <b>Warning:</b> the resulting index may be quite large.
Default=<code>trypsin</code>.</li>

<li>
<code>--digestion &lt;full-digest|partial-digest&gt;</code> &ndash; Specify
whether every peptide in the database must have two enzymatic termini
(<code>full-digest</code>) or if peptides with only one enzymatic
terminus are also included (<code>partial-digest</code>).
Default=<code>full-digest</code></li>

<li>
<code>--missed-cleavages &lt;integer&gt;</code> &ndash; Maximum number of missed
cleavages per peptide to allow in enzymatic digestion.  If this option is not
specified, then missed cleavages are not allowed.</li>

<li>
<code>--max-length &lt;integer&gt;</code> &ndash; Specify the maximum
length of the peptides included in the index.  Default=50.</li>

<li>
<code>--max-mass &lt;integer&gt;</code> &ndash; Specify the maximum
monoisotopic mass (in Da) of the peptides included in the index.
Default=7200.</li>

<li>
<code>--min-length &lt;integer&gt;</code> &ndash; Specify the minimum
length of the peptides included in the index.  Default=6.</li>

<li>
<code>--min-mass &lt;integer&gt;</code> &ndash; Specify the minimum
monoisotopic mass (in Da) of the peptides included in the index.
Default=200.</li>

<li>
<code>--monoisotopic-precursor &lt;T|F&gt;</code> &ndash; Use monoisotopic
precursor masses rather than average mass (default) for
precursor.
</li>

<li>
<code>--mods-spec &lt;string&gt;</code> &ndash; Expression for static and variable
      mass modifications to include. Specify a comma-separated list of
      modification specifications of the form:

      <p><code>--mods-spec C+57.02146,2M+15.9949,1STY+79.966331</code>,...

      <p>The general form of a modification specification has 3
      components, as exemplified
      by <code><red>1</red><green>STY</green>+<blue>79.966331</blue></code>.

      <p>The three components are:

      <p><code>[<red>max_per_peptide</red>]<green>residues</green>+<blue>added_mass</blue></code>.

      <p>In the example, <code><red>max_per_peptide</red></code>
      is <code><red>1</red></code>, <code><green>residues</green></code>
      are <code><green>STY</green></code>,
      and <code><blue>added_mass</blue></code>
      is <code><blue>79.966331</blue></code>.

<ul>
      <li><code><green>residues</green></code> indicate which amino
      acids may be affected by the modification.

      <li><code><blue>added_mass</blue></code> indicates in Daltons
      the mass to add to modified versions of the residues.

      <li><code><red>max_per_peptide</red></code> indicates how
      many occurrences of the indicated variable modification may
      occur per peptide. The <code><red>max_per_peptide</red></code>
      may be omitted, as in <code>C+57.02146</code>, to indicate a
      static modification, which will be in force at all occurrences
      of the amino acid residue.
</ul>

      <p>Notes: Currently only positive mass modifications are
      specifiable (<code><blue>added_mass</blue></code> &gt; 0). Use
      caution in specifying variable modifications: specifying many
      modifications or higher numbers of max_per_peptide can increase
      index size significantly.

</li>

<li>
<code>--output-dir &lt;filename&gt;</code> &ndash; 
The name of the directory where output files will be created.
Default = crux-output.</li>

<li>
<code>--overwrite &lt;T|F&gt;</code> &ndash;
Replace existing files if true (T) or fail when trying to overwrite a
file if false (F). Default =  F.</li>

<LI><code>--parameter-file &lt;filename&gt;</code> &ndash; 
A file containing command-line or additional parameters.  See
the <a href="parameter-file.html">parameter documentation</a> page
for details.  Default = no parameter file.</li>

<LI>
<code>--verbosity &lt;0-100&gt;</code> &ndash; 
Specify the verbosity of the current processes. Each level prints the
following messages, including all those at lower verbosity levels:
0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on
the progress of execution, 40-more progress information, 50-debug
info, 60-detailed debug info. Default = 30.</li>

</ul>

<hr></hr>

<p>
<a href="crux.html">Home</a></p>

</blockquote>
</body>
</html>


